

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>API Reference &mdash; sig2dna_core 0.43 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=b76d8e31"></script>
      <script src="_static/doctools.js?v=888ff710"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="📡🧬 sig2dna" href="README.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            sig2dna_core
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="README.html">📡🧬 sig2dna</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">API Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#module-sig2dna-core-signomics-py-from-the-generative-simulation-initiative">Module: sig2dna_core.signomics.py (from the Generative Simulation Initiative)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#main-components">Main Components</a></li>
<li class="toctree-l3"><a class="reference internal" href="#key-features">Key Features</a></li>
<li class="toctree-l3"><a class="reference internal" href="#core-concept">Core Concept</a></li>
<li class="toctree-l3"><a class="reference internal" href="#installation">Installation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#maintenance-forking">Maintenance &amp; forking</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sig2dna_core.signomics.DNACodes"><code class="docutils literal notranslate"><span class="pre">DNACodes</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNACodes.meta"><code class="docutils literal notranslate"><span class="pre">DNACodes.meta</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNACodes.encoded"><code class="docutils literal notranslate"><span class="pre">DNACodes.encoded</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNACodes.sinencode"><code class="docutils literal notranslate"><span class="pre">DNACodes.sinencode()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNACodes.sindecode"><code class="docutils literal notranslate"><span class="pre">DNACodes.sindecode()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNACodes.summary"><code class="docutils literal notranslate"><span class="pre">DNACodes.summary()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNACodes.plot"><code class="docutils literal notranslate"><span class="pre">DNACodes.plot()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id0"><code class="docutils literal notranslate"><span class="pre">DNACodes.plot()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id1"><code class="docutils literal notranslate"><span class="pre">DNACodes.sindecode()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id2"><code class="docutils literal notranslate"><span class="pre">DNACodes.sinencode()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id3"><code class="docutils literal notranslate"><span class="pre">DNACodes.summary()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sig2dna_core.signomics.DNAFullCodes"><code class="docutils literal notranslate"><span class="pre">DNAFullCodes</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAFullCodes.meta"><code class="docutils literal notranslate"><span class="pre">DNAFullCodes.meta</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAFullCodes.encoded"><code class="docutils literal notranslate"><span class="pre">DNAFullCodes.encoded</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAFullCodes.plot"><code class="docutils literal notranslate"><span class="pre">DNAFullCodes.plot</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id4"><code class="docutils literal notranslate"><span class="pre">DNAFullCodes.plot()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAFullCodes.sindecode"><code class="docutils literal notranslate"><span class="pre">DNAFullCodes.sindecode()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAFullCodes.sinencode"><code class="docutils literal notranslate"><span class="pre">DNAFullCodes.sinencode()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAFullCodes.summary"><code class="docutils literal notranslate"><span class="pre">DNAFullCodes.summary()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sig2dna_core.signomics.DNApairwiseAnalysis"><code class="docutils literal notranslate"><span class="pre">DNApairwiseAnalysis</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNApairwiseAnalysis.D"><code class="docutils literal notranslate"><span class="pre">DNApairwiseAnalysis.D</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNApairwiseAnalysis.names"><code class="docutils literal notranslate"><span class="pre">DNApairwiseAnalysis.names</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNApairwiseAnalysis.DNAsignals"><code class="docutils literal notranslate"><span class="pre">DNApairwiseAnalysis.DNAsignals</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNApairwiseAnalysis.coords"><code class="docutils literal notranslate"><span class="pre">DNApairwiseAnalysis.coords</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNApairwiseAnalysis.dimensions"><code class="docutils literal notranslate"><span class="pre">DNApairwiseAnalysis.dimensions</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNApairwiseAnalysis.linkage_matrix"><code class="docutils literal notranslate"><span class="pre">DNApairwiseAnalysis.linkage_matrix</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNApairwiseAnalysis.best_dimension"><code class="docutils literal notranslate"><span class="pre">DNApairwiseAnalysis.best_dimension()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNApairwiseAnalysis.cluster"><code class="docutils literal notranslate"><span class="pre">DNApairwiseAnalysis.cluster()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNApairwiseAnalysis.compute_linkage"><code class="docutils literal notranslate"><span class="pre">DNApairwiseAnalysis.compute_linkage()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNApairwiseAnalysis.dimension_variance_curve"><code class="docutils literal notranslate"><span class="pre">DNApairwiseAnalysis.dimension_variance_curve()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNApairwiseAnalysis.get_cluster_labels"><code class="docutils literal notranslate"><span class="pre">DNApairwiseAnalysis.get_cluster_labels()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNApairwiseAnalysis.heatmap"><code class="docutils literal notranslate"><span class="pre">DNApairwiseAnalysis.heatmap()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNApairwiseAnalysis.load"><code class="docutils literal notranslate"><span class="pre">DNApairwiseAnalysis.load()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNApairwiseAnalysis.pcoa"><code class="docutils literal notranslate"><span class="pre">DNApairwiseAnalysis.pcoa()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNApairwiseAnalysis.plot_dendrogram"><code class="docutils literal notranslate"><span class="pre">DNApairwiseAnalysis.plot_dendrogram()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNApairwiseAnalysis.reduced_distances"><code class="docutils literal notranslate"><span class="pre">DNApairwiseAnalysis.reduced_distances()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNApairwiseAnalysis.save"><code class="docutils literal notranslate"><span class="pre">DNApairwiseAnalysis.save()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNApairwiseAnalysis.scatter"><code class="docutils literal notranslate"><span class="pre">DNApairwiseAnalysis.scatter()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNApairwiseAnalysis.scatter3d"><code class="docutils literal notranslate"><span class="pre">DNApairwiseAnalysis.scatter3d()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNApairwiseAnalysis.select_dimensions"><code class="docutils literal notranslate"><span class="pre">DNApairwiseAnalysis.select_dimensions()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal"><code class="docutils literal notranslate"><span class="pre">DNAsignal</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.signal"><code class="docutils literal notranslate"><span class="pre">DNAsignal.signal</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.dtype"><code class="docutils literal notranslate"><span class="pre">DNAsignal.dtype</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.sampling_dt"><code class="docutils literal notranslate"><span class="pre">DNAsignal.sampling_dt</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.dx"><code class="docutils literal notranslate"><span class="pre">DNAsignal.dx</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.n"><code class="docutils literal notranslate"><span class="pre">DNAsignal.n</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.name"><code class="docutils literal notranslate"><span class="pre">DNAsignal.name</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.x_label"><code class="docutils literal notranslate"><span class="pre">DNAsignal.x_label</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.x_unit"><code class="docutils literal notranslate"><span class="pre">DNAsignal.x_unit</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.y_label"><code class="docutils literal notranslate"><span class="pre">DNAsignal.y_label</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.y_unit"><code class="docutils literal notranslate"><span class="pre">DNAsignal.y_unit</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.scales"><code class="docutils literal notranslate"><span class="pre">DNAsignal.scales</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.codes"><code class="docutils literal notranslate"><span class="pre">DNAsignal.codes</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.codesfull"><code class="docutils literal notranslate"><span class="pre">DNAsignal.codesfull</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id5"><code class="docutils literal notranslate"><span class="pre">DNAsignal.scales</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.transforms"><code class="docutils literal notranslate"><span class="pre">DNAsignal.transforms</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id6"><code class="docutils literal notranslate"><span class="pre">DNAsignal.codes</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id7"><code class="docutils literal notranslate"><span class="pre">DNAsignal.codesfull</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.peaks"><code class="docutils literal notranslate"><span class="pre">DNAsignal.peaks</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.codebook"><code class="docutils literal notranslate"><span class="pre">DNAsignal.codebook</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.generator"><code class="docutils literal notranslate"><span class="pre">DNAsignal.generator</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.compute_cwt"><code class="docutils literal notranslate"><span class="pre">DNAsignal.compute_cwt()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.encode_dna"><code class="docutils literal notranslate"><span class="pre">DNAsignal.encode_dna()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.encode_dna_full"><code class="docutils literal notranslate"><span class="pre">DNAsignal.encode_dna_full()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.plot_codes"><code class="docutils literal notranslate"><span class="pre">DNAsignal.plot_codes()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.decode_dna"><code class="docutils literal notranslate"><span class="pre">DNAsignal.decode_dna()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.__getitem__"><code class="docutils literal notranslate"><span class="pre">DNAsignal.__getitem__()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.summary"><code class="docutils literal notranslate"><span class="pre">DNAsignal.summary()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.has"><code class="docutils literal notranslate"><span class="pre">DNAsignal.has()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal._pairwiseEntropyDistance"><code class="docutils literal notranslate"><span class="pre">DNAsignal._pairwiseEntropyDistance()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal._pairwiseJaccardMotifDistance"><code class="docutils literal notranslate"><span class="pre">DNAsignal._pairwiseJaccardMotifDistance()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal._pairwiseJensenShannonDistance"><code class="docutils literal notranslate"><span class="pre">DNAsignal._pairwiseJensenShannonDistance()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal._pairwiseLevenshteinDistance"><code class="docutils literal notranslate"><span class="pre">DNAsignal._pairwiseLevenshteinDistance()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id8"><code class="docutils literal notranslate"><span class="pre">DNAsignal.encode_dna()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id9"><code class="docutils literal notranslate"><span class="pre">DNAsignal.encode_dna_full()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.sinencode_dna"><code class="docutils literal notranslate"><span class="pre">DNAsignal.sinencode_dna()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.sinencode_dna_full"><code class="docutils literal notranslate"><span class="pre">DNAsignal.sinencode_dna_full()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.sindecode_dna"><code class="docutils literal notranslate"><span class="pre">DNAsignal.sindecode_dna()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.sindecode_dna_full"><code class="docutils literal notranslate"><span class="pre">DNAsignal.sindecode_dna_full()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.tosignal"><code class="docutils literal notranslate"><span class="pre">DNAsignal.tosignal()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.align_with"><code class="docutils literal notranslate"><span class="pre">DNAsignal.align_with()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.apply_baseline_filter"><code class="docutils literal notranslate"><span class="pre">DNAsignal.apply_baseline_filter()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id10"><code class="docutils literal notranslate"><span class="pre">DNAsignal.compute_cwt()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id11"><code class="docutils literal notranslate"><span class="pre">DNAsignal.encode_dna()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id12"><code class="docutils literal notranslate"><span class="pre">DNAsignal.encode_dna_full()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.entropy_from_string"><code class="docutils literal notranslate"><span class="pre">DNAsignal.entropy_from_string()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.find_sequence"><code class="docutils literal notranslate"><span class="pre">DNAsignal.find_sequence()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.get_code"><code class="docutils literal notranslate"><span class="pre">DNAsignal.get_code()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.get_entropy"><code class="docutils literal notranslate"><span class="pre">DNAsignal.get_entropy()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id13"><code class="docutils literal notranslate"><span class="pre">DNAsignal.has()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id14"><code class="docutils literal notranslate"><span class="pre">DNAsignal.plot_codes()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.plot_signals"><code class="docutils literal notranslate"><span class="pre">DNAsignal.plot_signals()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.plot_transforms"><code class="docutils literal notranslate"><span class="pre">DNAsignal.plot_transforms()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.print_alignment"><code class="docutils literal notranslate"><span class="pre">DNAsignal.print_alignment()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.reconstruct_aligned_string"><code class="docutils literal notranslate"><span class="pre">DNAsignal.reconstruct_aligned_string()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.reconstruct_signal"><code class="docutils literal notranslate"><span class="pre">DNAsignal.reconstruct_signal()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id15"><code class="docutils literal notranslate"><span class="pre">DNAsignal.sindecode_dna()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id16"><code class="docutils literal notranslate"><span class="pre">DNAsignal.sinencode_dna()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id17"><code class="docutils literal notranslate"><span class="pre">DNAsignal.sinencode_dna_full()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal.synthetic_signal"><code class="docutils literal notranslate"><span class="pre">DNAsignal.synthetic_signal()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id18"><code class="docutils literal notranslate"><span class="pre">DNAsignal.tosignal()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr"><code class="docutils literal notranslate"><span class="pre">DNAstr</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.dx"><code class="docutils literal notranslate"><span class="pre">DNAstr.dx</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.iloc"><code class="docutils literal notranslate"><span class="pre">DNAstr.iloc</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.xloc"><code class="docutils literal notranslate"><span class="pre">DNAstr.xloc</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.aligned_with"><code class="docutils literal notranslate"><span class="pre">DNAstr.aligned_with</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.other_copy"><code class="docutils literal notranslate"><span class="pre">DNAstr.other_copy</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.ref_hash"><code class="docutils literal notranslate"><span class="pre">DNAstr.ref_hash</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.mask"><code class="docutils literal notranslate"><span class="pre">DNAstr.mask</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.engine"><code class="docutils literal notranslate"><span class="pre">DNAstr.engine</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.engineOpts"><code class="docutils literal notranslate"><span class="pre">DNAstr.engineOpts</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.align"><code class="docutils literal notranslate"><span class="pre">DNAstr.align()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.aligned_code"><code class="docutils literal notranslate"><span class="pre">DNAstr.aligned_code</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.alignment_stats"><code class="docutils literal notranslate"><span class="pre">DNAstr.alignment_stats</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.entropy"><code class="docutils literal notranslate"><span class="pre">DNAstr.entropy</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.excess_entropy"><code class="docutils literal notranslate"><span class="pre">DNAstr.excess_entropy()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.extract_motifs"><code class="docutils literal notranslate"><span class="pre">DNAstr.extract_motifs()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.find"><code class="docutils literal notranslate"><span class="pre">DNAstr.find()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.html_alignment"><code class="docutils literal notranslate"><span class="pre">DNAstr.html_alignment()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.jaccard"><code class="docutils literal notranslate"><span class="pre">DNAstr.jaccard()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.jensen_shannon"><code class="docutils literal notranslate"><span class="pre">DNAstr.jensen_shannon()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.levenshtein"><code class="docutils literal notranslate"><span class="pre">DNAstr.levenshtein()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.mutation_counts"><code class="docutils literal notranslate"><span class="pre">DNAstr.mutation_counts</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.mutual_entropy"><code class="docutils literal notranslate"><span class="pre">DNAstr.mutual_entropy()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.plot_alignment"><code class="docutils literal notranslate"><span class="pre">DNAstr.plot_alignment()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.plot_mask"><code class="docutils literal notranslate"><span class="pre">DNAstr.plot_mask()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.score"><code class="docutils literal notranslate"><span class="pre">DNAstr.score()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.summary"><code class="docutils literal notranslate"><span class="pre">DNAstr.summary()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.to_signal"><code class="docutils literal notranslate"><span class="pre">DNAstr.to_signal()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.vectorized"><code class="docutils literal notranslate"><span class="pre">DNAstr.vectorized()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.DNAstr.wrapped_alignment"><code class="docutils literal notranslate"><span class="pre">DNAstr.wrapped_alignment()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sig2dna_core.signomics.SinusoidalEncoder"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.SinusoidalEncoder.d_model"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.d_model</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.SinusoidalEncoder.N"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.N</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.SinusoidalEncoder.dtype"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.dtype</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.SinusoidalEncoder._last_input_type"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder._last_input_type</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.SinusoidalEncoder._last_input_length"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder._last_input_length</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.SinusoidalEncoder._scale"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder._scale</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.SinusoidalEncoder._auto_scale_enabled"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder._auto_scale_enabled</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.SinusoidalEncoder._decode_residual_tolerance"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder._decode_residual_tolerance</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.SinusoidalEncoder.encode"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.encode()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.SinusoidalEncoder.decode"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.decode()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.SinusoidalEncoder.fit_encoder"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.fit_encoder()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.SinusoidalEncoder.set_decode_tolerance"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.set_decode_tolerance()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.SinusoidalEncoder.verify_roundtrip"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.verify_roundtrip()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.SinusoidalEncoder.sinencode_dna_grouped"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.sinencode_dna_grouped()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.SinusoidalEncoder.sindecode_dna_grouped"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.sindecode_dna_grouped()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.SinusoidalEncoder.sinencode_dnafull_grouped"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.sinencode_dnafull_grouped()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.SinusoidalEncoder.sindecode_dnafull_grouped"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.sindecode_dnafull_grouped()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.SinusoidalEncoder.to_complex"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.to_complex()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.SinusoidalEncoder.complex_distance"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.complex_distance()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.SinusoidalEncoder.angle_difference"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.angle_difference()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.SinusoidalEncoder.phase_alignment"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.phase_alignment()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.SinusoidalEncoder.pairwise_similarity"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.pairwise_similarity()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.SinusoidalEncoder.group_centroid"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.group_centroid()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.SinusoidalEncoder.phase_unwrap"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.phase_unwrap()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.SinusoidalEncoder.Example"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.Example()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id19"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.Example()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.SinusoidalEncoder.References"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.References()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id20"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.angle_difference()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id21"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.complex_distance()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id22"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.decode()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id23"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.encode()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id24"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.fit_encoder()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id25"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.group_centroid()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id26"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.pairwise_similarity()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id27"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.phase_alignment()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id28"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.phase_unwrap()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id29"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.set_decode_tolerance()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id30"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.sindecode_dna_grouped()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id31"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.sinencode_dna_grouped()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id32"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.to_complex()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id33"><code class="docutils literal notranslate"><span class="pre">SinusoidalEncoder.verify_roundtrip()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sig2dna_core.signomics.generator"><code class="docutils literal notranslate"><span class="pre">generator</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#sig2dna_core.signomics.import_local_module"><code class="docutils literal notranslate"><span class="pre">import_local_module()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#sig2dna_core.signomics.peaks"><code class="docutils literal notranslate"><span class="pre">peaks</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.peaks.add"><code class="docutils literal notranslate"><span class="pre">peaks.add()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.peaks.as_dict"><code class="docutils literal notranslate"><span class="pre">peaks.as_dict()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.peaks.copy"><code class="docutils literal notranslate"><span class="pre">peaks.copy()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.peaks.names"><code class="docutils literal notranslate"><span class="pre">peaks.names()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.peaks.remove_duplicates"><code class="docutils literal notranslate"><span class="pre">peaks.remove_duplicates()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.peaks.rename"><code class="docutils literal notranslate"><span class="pre">peaks.rename()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.peaks.sort"><code class="docutils literal notranslate"><span class="pre">peaks.sort()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.peaks.to_signal"><code class="docutils literal notranslate"><span class="pre">peaks.to_signal()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.peaks.update"><code class="docutils literal notranslate"><span class="pre">peaks.update()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sig2dna_core.signomics.signal"><code class="docutils literal notranslate"><span class="pre">signal</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal.x"><code class="docutils literal notranslate"><span class="pre">signal.x</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal.y"><code class="docutils literal notranslate"><span class="pre">signal.y</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal.name"><code class="docutils literal notranslate"><span class="pre">signal.name</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal.type"><code class="docutils literal notranslate"><span class="pre">signal.type</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal.x_label"><code class="docutils literal notranslate"><span class="pre">signal.x_label</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal.x_unit"><code class="docutils literal notranslate"><span class="pre">signal.x_unit</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal.source"><code class="docutils literal notranslate"><span class="pre">signal.source</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal.metadata"><code class="docutils literal notranslate"><span class="pre">signal.metadata</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal.color"><code class="docutils literal notranslate"><span class="pre">signal.color</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal._previous"><code class="docutils literal notranslate"><span class="pre">signal._previous</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal._history"><code class="docutils literal notranslate"><span class="pre">signal._history</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal.add_noise"><code class="docutils literal notranslate"><span class="pre">signal.add_noise()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal.align_with"><code class="docutils literal notranslate"><span class="pre">signal.align_with()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal.apply_poisson_baseline_filter"><code class="docutils literal notranslate"><span class="pre">signal.apply_poisson_baseline_filter()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal.backup"><code class="docutils literal notranslate"><span class="pre">signal.backup()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal.copy"><code class="docutils literal notranslate"><span class="pre">signal.copy()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal.disable_fullhistory"><code class="docutils literal notranslate"><span class="pre">signal.disable_fullhistory()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal.enable_fullhistory"><code class="docutils literal notranslate"><span class="pre">signal.enable_fullhistory()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal.from_peaks"><code class="docutils literal notranslate"><span class="pre">signal.from_peaks()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal.load"><code class="docutils literal notranslate"><span class="pre">signal.load()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal.n"><code class="docutils literal notranslate"><span class="pre">signal.n</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal.plot"><code class="docutils literal notranslate"><span class="pre">signal.plot()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal.restore"><code class="docutils literal notranslate"><span class="pre">signal.restore()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal.sample"><code class="docutils literal notranslate"><span class="pre">signal.sample()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal.save"><code class="docutils literal notranslate"><span class="pre">signal.save()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sig2dna_core.signomics.signal_collection"><code class="docutils literal notranslate"><span class="pre">signal_collection</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal_collection.append"><code class="docutils literal notranslate"><span class="pre">signal_collection.append()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal_collection.generate_mixtures"><code class="docutils literal notranslate"><span class="pre">signal_collection.generate_mixtures()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal_collection.generate_synthetic"><code class="docutils literal notranslate"><span class="pre">signal_collection.generate_synthetic()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal_collection.mean"><code class="docutils literal notranslate"><span class="pre">signal_collection.mean()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal_collection.plot"><code class="docutils literal notranslate"><span class="pre">signal_collection.plot()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal_collection.sum"><code class="docutils literal notranslate"><span class="pre">signal_collection.sum()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#sig2dna_core.signomics.signal_collection.to_matrix"><code class="docutils literal notranslate"><span class="pre">signal_collection.to_matrix()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">sig2dna_core</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">API Reference</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/api.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-sig2dna_core.signomics">
<span id="api-reference"></span><h1>API Reference<a class="headerlink" href="#module-sig2dna_core.signomics" title="Link to this heading"></a></h1>
<section id="module-sig2dna-core-signomics-py-from-the-generative-simulation-initiative">
<h2>Module: sig2dna_core.signomics.py (from the Generative Simulation Initiative)<a class="headerlink" href="#module-sig2dna-core-signomics-py-from-the-generative-simulation-initiative" title="Link to this heading"></a></h2>
<p>This is the core module of the sig2dna framework, dedicated to transforming numerical chemical signals into DNA-like symbolic representations. The module enables symbolic analysis, fingerprinting, alignment, and classification of complex analytical signals, such as:</p>
<ul class="simple">
<li><p>GC-MS / GC-FID</p></li>
<li><p>HPLC-MS</p></li>
<li><p>NMR / FTIR / Raman</p></li>
<li><p>RX and other spectroscopy data</p></li>
</ul>
<p>It is designed to facilitate high-throughput pattern recognition, compression, clustering, and AI/ML-based classification. Symbolic transformation is based on <strong>wavelet decomposition (Mexican Hat/Ricker)</strong> and segment encoding into letters (e.g., A, B, C, X, Y, Z). The representation preserves key structural patterns (e.g., peak transitions) across multiple scales and supports entropy-based distances.</p>
<section id="main-components">
<h3>Main Components<a class="headerlink" href="#main-components" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>DNAsignal — core class to transform a signal into DNA-like symbolic representation</p></li>
<li><p>DNAstr — string subclass enabling alignment, entropy analysis, visualization, and reconstruction</p></li>
<li><p>DNApairwiseAnalysis — distance and clustering toolbox for aligned DNA codes (PCoA, dendrogram, 2D/3D plots)</p></li>
</ul>
</section>
<section id="key-features">
<h3>Key Features<a class="headerlink" href="#key-features" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>Multi-scale wavelet transform with symbolic encoding</p></li>
<li><p>Symbolic entropy and mutual information measures</p></li>
<li><p>Fast symbolic alignment using <cite>difflib</cite> or <cite>biopython</cite></p></li>
<li><p>Pairwise symbolic distances (Shannon, excess entropy, Jaccard, Jensen-Shannon)</p></li>
<li><p>Interactive plotting: segments, alignment masks, triangle patches</p></li>
<li><p>Motif search, alignment visualization, HTML and terminal rendering</p></li>
<li><p>Dimensionality reduction (MDS), clustering, dendrogram and heatmaps</p></li>
</ul>
</section>
<section id="core-concept">
<h3>Core Concept<a class="headerlink" href="#core-concept" title="Link to this heading"></a></h3>
<p>Input Signal
- 1D NumPy array S of shape (m,)
- Data type: np.float64 (default) or np.float32
- Typically sparse and non-negative, such as GC-MS total ion chromatograms</p>
<p>Wavelet Transform
- CWT with Ricker (Mexican hat) wavelet
- Scales: $s = 2^0, 2^1,…, 2^n
- Downsampling by scale (to reduce data volume and capture features at relevant resolutions)</p>
<p>Symbolic Encoding and compressed representation</p>
<p>The transformed signal Ts at scale s is converted to a sequence of symbolic letters using the rules:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Symbol</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>A</p></td>
<td><p>Monotonic increase crossing from − to +</p></td>
</tr>
<tr class="row-odd"><td><p>B</p></td>
<td><p>Monotonic increase from − to − (no zero crossing)</p></td>
</tr>
<tr class="row-even"><td><p>C</p></td>
<td><p>Monotonic increase from + to +</p></td>
</tr>
<tr class="row-odd"><td><p>X</p></td>
<td><p>Monotonic decrease from + to + (no zero crossing)</p></td>
</tr>
<tr class="row-even"><td><p>Y</p></td>
<td><p>Monotonic decrease from − to −</p></td>
</tr>
<tr class="row-odd"><td><p>Z</p></td>
<td><p>Monotonic decrease crossing from + to −</p></td>
</tr>
<tr class="row-even"><td><p>_</p></td>
<td><p>Zero or noise (after filtering)</p></td>
</tr>
</tbody>
</table>
<p>Each encoded segment is associated with:
- width: number of points
- height: amplitude difference</p>
<p>These form the compressed representation</p>
</section>
<section id="installation">
<h3>Installation<a class="headerlink" href="#installation" title="Link to this heading"></a></h3>
<p>Install all dependencies with:</p>
<p>conda install pywavelets seaborn scikit-learn
conda install -c conda-forge python-Levenshtein biopython</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">signomics</span><span class="w"> </span><span class="kn">import</span> <span class="n">DNAsignal</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">signal</span><span class="w"> </span><span class="kn">import</span> <span class="n">signal</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># Load a sampled signal (e.g., from GC-MS, Raman)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">S</span> <span class="o">=</span> <span class="n">signal</span><span class="o">.</span><span class="n">from_peaks</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>  <span class="c1"># or any constructor for sampled signals</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># Encode into DNA-like format</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">D</span> <span class="o">=</span> <span class="n">DNAsignal</span><span class="p">(</span><span class="n">S</span><span class="p">,</span> <span class="n">encode</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">D</span><span class="o">.</span><span class="n">encode_dna</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">D</span><span class="o">.</span><span class="n">plot_codes</span><span class="p">(</span><span class="n">scale</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># Compare samples and cluster</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dlist</span> <span class="o">=</span> <span class="p">[</span><span class="n">DNAsignal</span><span class="p">(</span><span class="n">S1</span><span class="p">,</span> <span class="n">encode</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span> <span class="n">DNAsignal</span><span class="p">(</span><span class="n">S2</span><span class="p">,</span> <span class="n">encode</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span> <span class="o">...</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">analysis</span> <span class="o">=</span> <span class="n">DNAsignal</span><span class="o">.</span><span class="n">_pairwiseEntropyDistance</span><span class="p">(</span><span class="n">Dlist</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">analysis</span><span class="o">.</span><span class="n">plot_dendrogram</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">analysis</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">n_clusters</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
<p class="rubric">Notes</p>
<p>The methodology implemented in this module covers and extends the approaches initially tested during the PhD of Julien Kermorvant.
“Concept of chemical fingerprints applied to the management of chemical risk of materials, recycled deposits and food packaging”. PhD thesis AgroParisTech. December 2023. <a class="reference external" href="https://theses.hal.science/tel-04194172">https://theses.hal.science/tel-04194172</a></p>
</section>
<section id="maintenance-forking">
<h3>Maintenance &amp; forking<a class="headerlink" href="#maintenance-forking" title="Link to this heading"></a></h3>
<blockquote>
<div><p>$ git init -b main
$ gh repo create sig2dna –public –source=. –remote=origin –push
$ # alternatively
$ # git remote add origin <a class="reference external" href="mailto:git&#37;&#52;&#48;github&#46;com">git<span>&#64;</span>github<span>&#46;</span>com</a>:ovitrac/sig2dna.git
$ # git branch -M main        # Ensure current branch is named ‘main’
$ # git push -u origin main   # Push and set upstream tracking</p>
<p>$ tree -P ‘<em>.py’ -P ‘</em>.md’ -P ‘LICENSE’ -I ‘__pycache__|.*’ –prune
$ pdoc ./sig2dna-core/signomics.py -f –html -o ./docs
$ doctoc –github  –maxlevel 2 README.md</p>
</div></blockquote>
<p>Author: Olivier Vitrac — <a class="reference external" href="mailto:olivier&#46;vitrac&#37;&#52;&#48;gmail&#46;com">olivier<span>&#46;</span>vitrac<span>&#64;</span>gmail<span>&#46;</span>com</a>
Revision: 2025-05-27</p>
</section>
</section>
<dl class="py class">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNACodes">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sig2dna_core.signomics.</span></span><span class="sig-name descname"><span class="pre">DNACodes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">encoded</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNACodes" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">UserDict</span></code></p>
<p>🧬 DNACodes
Dictionary-like container for symbolic signal encodings at multiple scales.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNACodes.meta">
<span class="sig-name descname"><span class="pre">meta</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNACodes.meta" title="Link to this definition"></a></dt>
<dd><p>Metadata describing the signal and encoding parameters.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNACodes.encoded">
<span class="sig-name descname"><span class="pre">encoded</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNACodes.encoded" title="Link to this definition"></a></dt>
<dd><p>Whether the content has been sinusoidally encoded.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNACodes.sinencode">
<span class="sig-name descname"><span class="pre">sinencode</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d_part</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">32</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">N</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNACodes.sinencode" title="Link to this definition"></a></dt>
<dd><p>Encodes symbolic segments using transformer-style sinusoidal embeddings.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNACodes.sindecode">
<span class="sig-name descname"><span class="pre">sindecode</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reference_dx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNACodes.sindecode" title="Link to this definition"></a></dt>
<dd><p>Decodes sinusoidal embeddings back to symbolic segment structure.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNACodes.summary">
<span class="sig-name descname"><span class="pre">summary</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNACodes.summary" title="Link to this definition"></a></dt>
<dd><p>Displays segment or vector counts by scale.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNACodes.plot">
<span class="sig-name descname"><span class="pre">plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(12,</span> <span class="pre">4)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">d_part</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">N</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNACodes.plot" title="Link to this definition"></a></dt>
<dd><p>Plot method for DNACodes</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id0">
<span class="sig-name descname"><span class="pre">plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(12,</span> <span class="pre">4)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">d_part</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">N</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id0" title="Link to this definition"></a></dt>
<dd><p>Plot method for DNACodes: visualizes encoded vectors or symbolic segment distribution.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>figsize</strong> (<em>tuple</em>) – Figure size for the entire plot.</p></li>
<li><p><strong>d_part</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of dimensions per segment part (only for encoded).</p></li>
<li><p><strong>N</strong> (<em>int</em><em>, </em><em>optional</em>) – Frequency base (for metadata or title info).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id1">
<span class="sig-name descname"><span class="pre">sindecode</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reference_dx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id1" title="Link to this definition"></a></dt>
<dd><p>Decode sinusoidally embedded codes grouped by letter into symbolic segment structure.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>reference_dx</strong> (<em>float</em><em>, </em><em>optional</em>) – Sampling interval used to reconstruct <cite>xloc</cite>. Defaults to meta[“sampling_dt”].</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Decoded symbolic codes for each scale.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sig2dna_core.signomics.DNACodes" title="sig2dna_core.signomics.DNACodes">DNACodes</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id2">
<span class="sig-name descname"><span class="pre">sinencode</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d_part</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">32</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">N</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id2" title="Link to this definition"></a></dt>
<dd><p>Encode symbolic segments at each scale using sinusoidal encoding grouped by letter.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>d_part</strong> (<em>int</em>) – Number of dimensions for each component (start, width, height).</p></li>
<li><p><strong>N</strong> (<em>int</em>) – Frequency base for sinusoidal embedding.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Encoded version of the current codes, grouped by letter per scale.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sig2dna_core.signomics.DNACodes" title="sig2dna_core.signomics.DNACodes">DNACodes</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id3">
<span class="sig-name descname"><span class="pre">summary</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#id3" title="Link to this definition"></a></dt>
<dd><p>Print the number of encoded vectors or symbolic segments per scale.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAFullCodes">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sig2dna_core.signomics.</span></span><span class="sig-name descname"><span class="pre">DNAFullCodes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">encoded</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAFullCodes" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></p>
<section id="dnafullcodes-dict">
<h2>🧬 DNAFullCodes(dict)<a class="headerlink" href="#dnafullcodes-dict" title="Link to this heading"></a></h2>
<p>Container for symbolic full-resolution DNA-like strings per scale.</p>
<p>Each entry in the dictionary maps a scale (int) to a DNAstr object or str sequence.
This class supports future extensions such as sinusoidal encoding or pattern-based decoding.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAFullCodes.meta">
<span class="sig-name descname"><span class="pre">meta</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAFullCodes.meta" title="Link to this definition"></a></dt>
<dd><p>Optional metadata about sampling, units, labels, etc.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAFullCodes.encoded">
<span class="sig-name descname"><span class="pre">encoded</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAFullCodes.encoded" title="Link to this definition"></a></dt>
<dd><p>True if this structure has been encoded via a sinusoidal encoder.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAFullCodes.plot">
<span class="sig-name descname"><span class="pre">plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(12,</span> <span class="pre">4)</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAFullCodes.plot" title="Link to this definition"></a></dt>
<dd><p>Plot method for DNAFullCodes</p>
</dd></dl>

<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">codes</span> <span class="o">=</span> <span class="n">DNAFullCodes</span><span class="p">({</span><span class="mi">4</span><span class="p">:</span> <span class="s1">&#39;YAABZZ&#39;</span><span class="p">},</span> <span class="n">meta</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;sampling_dt&quot;</span><span class="p">:</span> <span class="mf">0.5</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">codes</span><span class="o">.</span><span class="n">summary</span><span class="p">())</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="id4">
<span class="sig-name descname"><span class="pre">plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(12,</span> <span class="pre">4)</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id4" title="Link to this definition"></a></dt>
<dd><p>Plot method for DNAFullCodes: visualizes encoded vectors or DNA string composition.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>figsize</strong> (<em>tuple</em>) – Figure size for the entire plot.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAFullCodes.sindecode">
<span class="sig-name descname"><span class="pre">sindecode</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAFullCodes.sindecode" title="Link to this definition"></a></dt>
<dd><p>Sindecode method</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAFullCodes.sinencode">
<span class="sig-name descname"><span class="pre">sinencode</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d_model</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">96</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">N</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAFullCodes.sinencode" title="Link to this definition"></a></dt>
<dd><p>Sinencode method</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAFullCodes.summary">
<span class="sig-name descname"><span class="pre">summary</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAFullCodes.summary" title="Link to this definition"></a></dt>
<dd><p>Return a brief summary of the full codes per scale.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Mapping from scale to summary string or code length.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
</dd></dl>

</section>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNApairwiseAnalysis">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sig2dna_core.signomics.</span></span><span class="sig-name descname"><span class="pre">DNApairwiseAnalysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">D</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">DNAsignals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNApairwiseAnalysis" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class to handle pairwise distance analysis, PCoA, clustering, and visualization
for DNA-coded signals.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNApairwiseAnalysis.D">
<span class="sig-name descname"><span class="pre">D</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNApairwiseAnalysis.D" title="Link to this definition"></a></dt>
<dd><p>Pairwise excess entropy distance matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNApairwiseAnalysis.names">
<span class="sig-name descname"><span class="pre">names</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNApairwiseAnalysis.names" title="Link to this definition"></a></dt>
<dd><p>Names of the DNA signals.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNApairwiseAnalysis.DNAsignals">
<span class="sig-name descname"><span class="pre">DNAsignals</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNApairwiseAnalysis.DNAsignals" title="Link to this definition"></a></dt>
<dd><p>original DNAsignal objects</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNApairwiseAnalysis.coords">
<span class="sig-name descname"><span class="pre">coords</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNApairwiseAnalysis.coords" title="Link to this definition"></a></dt>
<dd><p>Coordinates in reduced space (PCoA).</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNApairwiseAnalysis.dimensions">
<span class="sig-name descname"><span class="pre">dimensions</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNApairwiseAnalysis.dimensions" title="Link to this definition"></a></dt>
<dd><p>Selected dimensions for reduced analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNApairwiseAnalysis.linkage_matrix">
<span class="sig-name descname"><span class="pre">linkage_matrix</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNApairwiseAnalysis.linkage_matrix" title="Link to this definition"></a></dt>
<dd><p>Linkage matrix used for hierarchical clustering.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNApairwiseAnalysis.best_dimension">
<span class="sig-name descname"><span class="pre">best_dimension</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">max_dim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNApairwiseAnalysis.best_dimension" title="Link to this definition"></a></dt>
<dd><p>Determine optimal dimension by maximizing silhouette score.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNApairwiseAnalysis.cluster">
<span class="sig-name descname"><span class="pre">cluster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">criterion</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'distance'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNApairwiseAnalysis.cluster" title="Link to this definition"></a></dt>
<dd><p>Assign cluster labels from linkage matrix.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNApairwiseAnalysis.compute_linkage">
<span class="sig-name descname"><span class="pre">compute_linkage</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ward'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNApairwiseAnalysis.compute_linkage" title="Link to this definition"></a></dt>
<dd><p>Compute hierarchical clustering.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNApairwiseAnalysis.dimension_variance_curve">
<span class="sig-name descname"><span class="pre">dimension_variance_curve</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(8,</span> <span class="pre">5)</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNApairwiseAnalysis.dimension_variance_curve" title="Link to this definition"></a></dt>
<dd><p>Computes the cumulative explained variance (based on pairwise distances) as a function of
the number of dimensions used (from 1 to n-1). Optionally plots the curve and the point
where the threshold (default 0.5) is reached.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>threshold</strong> (<em>float</em>) – Fraction of total variance to reach (default 0.5).</p></li>
<li><p><strong>plot</strong> (<em>bool</em>) – If True, display the variance curve and highlight dhalf.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em>) – Size of the figure if plotted.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>dhalf</strong> (<em>int</em>) – Number of dimensions needed to reach the threshold.</p></li>
<li><p><strong>curve</strong> (<em>list of float</em>) – Normalized cumulative variance (in [0, 1]) for dimensions 1 to n-1.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNApairwiseAnalysis.get_cluster_labels">
<span class="sig-name descname"><span class="pre">get_cluster_labels</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_clusters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ward'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNApairwiseAnalysis.get_cluster_labels" title="Link to this definition"></a></dt>
<dd><p>Returns cluster labels from hierarchical clustering. If not computed yet, computes linkage.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n_clusters</strong> (<em>int</em>) – Number of clusters to assign.</p></li>
<li><p><strong>method</strong> (<em>str</em>) – Linkage method to use if recomputing linkage.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>labels</strong> – Cluster IDs for each sample.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray of int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNApairwiseAnalysis.heatmap">
<span class="sig-name descname"><span class="pre">heatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(10,</span> <span class="pre">8)</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNApairwiseAnalysis.heatmap" title="Link to this definition"></a></dt>
<dd><p>Plot heatmap of pairwise distances.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNApairwiseAnalysis.load">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNApairwiseAnalysis.load" title="Link to this definition"></a></dt>
<dd><p>Load analysis from file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNApairwiseAnalysis.pcoa">
<span class="sig-name descname"><span class="pre">pcoa</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_components</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNApairwiseAnalysis.pcoa" title="Link to this definition"></a></dt>
<dd><p>Perform Principal Coordinate Analysis (PCoA).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNApairwiseAnalysis.plot_dendrogram">
<span class="sig-name descname"><span class="pre">plot_dendrogram</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">truncate_mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNApairwiseAnalysis.plot_dendrogram" title="Link to this definition"></a></dt>
<dd><p>Plot dendrogram from linkage matrix.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNApairwiseAnalysis.reduced_distances">
<span class="sig-name descname"><span class="pre">reduced_distances</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNApairwiseAnalysis.reduced_distances" title="Link to this definition"></a></dt>
<dd><p>Recompute distances on selected subspace.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNApairwiseAnalysis.save">
<span class="sig-name descname"><span class="pre">save</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNApairwiseAnalysis.save" title="Link to this definition"></a></dt>
<dd><p>Save current analysis to file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNApairwiseAnalysis.scatter">
<span class="sig-name descname"><span class="pre">scatter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dims</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0,</span> <span class="pre">1)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(8,</span> <span class="pre">6)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_clusters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNApairwiseAnalysis.scatter" title="Link to this definition"></a></dt>
<dd><p>2D scatter plot in selected dimensions with optional cluster-based coloring.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dims</strong> (<em>tuple</em>) – Dimensions to plot (default: (0, 1)).</p></li>
<li><p><strong>annotate</strong> (<em>bool</em>) – If True, annotate points with their index.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em>) – Size of the plot.</p></li>
<li><p><strong>n_clusters</strong> (<em>int</em><em> or </em><em>None</em>) – If provided, use clustering to color points.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>fig</strong></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>matplotlib.figure.Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNApairwiseAnalysis.scatter3d">
<span class="sig-name descname"><span class="pre">scatter3d</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dims</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0,</span> <span class="pre">1,</span> <span class="pre">2)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_clusters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNApairwiseAnalysis.scatter3d" title="Link to this definition"></a></dt>
<dd><p>3D scatter plot in selected dimensions with optional cluster-based coloring.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dims</strong> (<em>tuple</em>) – Dimensions to plot.</p></li>
<li><p><strong>annotate</strong> (<em>bool</em>) – Annotate points with their index.</p></li>
<li><p><strong>n_clusters</strong> (<em>int</em><em> or </em><em>None</em>) – If provided, use clustering to color points.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>fig</strong></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>matplotlib.figure.Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNApairwiseAnalysis.select_dimensions">
<span class="sig-name descname"><span class="pre">select_dimensions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dims</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNApairwiseAnalysis.select_dimensions" title="Link to this definition"></a></dt>
<dd><p>Update active dimensions.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sig2dna_core.signomics.</span></span><span class="sig-name descname"><span class="pre">DNAsignal</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">signal_obj,</span> <span class="pre">sampling_dt=1.0,</span> <span class="pre">dtype=&lt;class</span> <span class="pre">'numpy.float64'&gt;,</span> <span class="pre">encode=False,</span> <span class="pre">encoder=['compute_cwt',</span> <span class="pre">'encode_dna',</span> <span class="pre">'encode_dna_full'],</span> <span class="pre">scales=[1,</span> <span class="pre">2,</span> <span class="pre">4,</span> <span class="pre">8,</span> <span class="pre">16,</span> <span class="pre">32],</span> <span class="pre">x_label='index',</span> <span class="pre">x_unit='-',</span> <span class="pre">y_label='Intensity',</span> <span class="pre">y_unit='',</span> <span class="pre">plot=False,</span> <span class="pre">plotter=['plot_signals',</span> <span class="pre">'plot_transforms',</span> <span class="pre">'plot_codes']</span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<section id="dnasignal-signal">
<h2>DNAsignal(signal):<a class="headerlink" href="#dnasignal-signal" title="Link to this heading"></a></h2>
<p>A class to encode a numerical signal (typically a 1D GC-MS trace, NMR/FTIR/Raman spectra
or time series) into a DNA-like symbolic representation using wavelet analysis.
This symbolic coding enables fast comparison, search, and alignment of signal features
using abstracted patterns (e.g., ‘YAZB’).</p>
<p>The class supports:
- Continuous Wavelet Transform (CWT) using Ricker wavelets
- Symbolic conversion of wavelet features to DNA-like letters (A, B, C, X, Y, Z)
- Visualization of CWTs, symbolic encodings, and signal overlays
- Reversible decoding of symbolic segments back into approximate signals
- Substring extraction and matching
- Storage of multi-scale representations (multi-resolution DNA encoding)</p>
<p>This class is part of the symbolic signal transformation pipeline (<cite>sig2dna</cite>), compatible with <cite>signal</cite>, <cite>peaks</cite>, and <cite>signal_collection</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">param signal<span class="colon">:</span></dt>
<dd class="field-odd"><p>An instance of the <cite>signal</cite> class representing a sampled waveform. It must
have <cite>x</cite>, <cite>y</cite>, and a valid name or identifier.</p>
</dd>
<dt class="field-even">type signal<span class="colon">:</span></dt>
<dd class="field-even"><p>signal</p>
</dd>
<dt class="field-odd">param encode<span class="colon">:</span></dt>
<dd class="field-odd"><p>Launch encoders = [“compute_cwt”,”encode_dna”,”encode_dna_full”] if True</p>
</dd>
<dt class="field-even">type encode<span class="colon">:</span></dt>
<dd class="field-even"><p>bool (default = False)</p>
</dd>
<dt class="field-odd">param plot<span class="colon">:</span></dt>
<dd class="field-odd"><p>Plot with plotter = [“plot_signals”,”plot_transforms”,”plot_codes”]</p>
</dd>
<dt class="field-even">type plot<span class="colon">:</span></dt>
<dd class="field-even"><p>bool (default = False)</p>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.signal">
<span class="sig-name descname"><span class="pre">signal</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.signal" title="Link to this definition"></a></dt>
<dd><p>Original numerical signal (values only; x stored via <cite>sampling_dt</cite>).</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.dtype" title="Link to this definition"></a></dt>
<dd><p>Data type of the stored signal.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>data-type</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.sampling_dt">
<span class="sig-name descname"><span class="pre">sampling_dt</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.sampling_dt" title="Link to this definition"></a></dt>
<dd><p>Sampling interval along the x-axis.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.dx">
<span class="sig-name descname"><span class="pre">dx</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.dx" title="Link to this definition"></a></dt>
<dd><p>The nominal x-resolution of the signal (automatically derived).</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float (depreciated)</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.n">
<span class="sig-name descname"><span class="pre">n</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.n" title="Link to this definition"></a></dt>
<dd><p>Number of points in the signal (length of x/y arrays).</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.name">
<span class="sig-name descname"><span class="pre">name</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.name" title="Link to this definition"></a></dt>
<dd><p>Name of the signal.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.x_label">
<span class="sig-name descname"><span class="pre">x_label</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.x_label" title="Link to this definition"></a></dt>
<dd><p>Label of the x-axis.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.x_unit">
<span class="sig-name descname"><span class="pre">x_unit</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.x_unit" title="Link to this definition"></a></dt>
<dd><p>Unit of the x-axis.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.y_label">
<span class="sig-name descname"><span class="pre">y_label</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.y_label" title="Link to this definition"></a></dt>
<dd><p>Label of the y-axis.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.y_unit">
<span class="sig-name descname"><span class="pre">y_unit</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.y_unit" title="Link to this definition"></a></dt>
<dd><p>Unit of the y-axis.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.scales">
<span class="sig-name descname"><span class="pre">scales</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.scales" title="Link to this definition"></a></dt>
<dd><p>List of scales used for encoding.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[int]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.codes">
<span class="sig-name descname"><span class="pre">codes</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.codes" title="Link to this definition"></a></dt>
<dd><p>Dictionary-like container of symbolic triplet encodings by scale.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sig2dna_core.signomics.DNACodes" title="sig2dna_core.signomics.DNACodes">DNACodes</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.codesfull">
<span class="sig-name descname"><span class="pre">codesfull</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.codesfull" title="Link to this definition"></a></dt>
<dd><p>Dictionary-like container of full-resolution symbolic strings per scale.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sig2dna_core.signomics.DNAFullCodes" title="sig2dna_core.signomics.DNAFullCodes">DNAFullCodes</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id5">
<span class="sig-name descname"><span class="pre">scales</span></span><a class="headerlink" href="#id5" title="Link to this definition"></a></dt>
<dd><p>Set of scales used in the Continuous Wavelet Transform (powers of 2 by default).</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>array-like</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.transforms">
<span class="sig-name descname"><span class="pre">transforms</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.transforms" title="Link to this definition"></a></dt>
<dd><p>Stores the CWT-transformed signals, each as a <cite>signal</cite> object, indexed by scale.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sig2dna_core.signomics.signal_collection" title="sig2dna_core.signomics.signal_collection">signal_collection</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id6">
<span class="sig-name descname"><span class="pre">codes</span></span><a class="headerlink" href="#id6" title="Link to this definition"></a></dt>
<dd><p>Symbolic codes by scale level. Each is a <cite>DNAstr</cite> object representing
the symbolic sequence at that scale.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[int, <a class="reference internal" href="#sig2dna_core.signomics.DNAstr" title="sig2dna_core.signomics.DNAstr">DNAstr</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id7">
<span class="sig-name descname"><span class="pre">codesfull</span></span><a class="headerlink" href="#id7" title="Link to this definition"></a></dt>
<dd><p>Same as <cite>codes</cite>, but uses full resolution symbolic representation.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[int, <a class="reference internal" href="#sig2dna_core.signomics.DNAstr" title="sig2dna_core.signomics.DNAstr">DNAstr</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.peaks">
<span class="sig-name descname"><span class="pre">peaks</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.peaks" title="Link to this definition"></a></dt>
<dd><p>Optional <cite>peaks</cite> object used to index real peak positions from the signal.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sig2dna_core.signomics.peaks" title="sig2dna_core.signomics.peaks">peaks</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.codebook">
<span class="sig-name descname"><span class="pre">codebook</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.codebook" title="Link to this definition"></a></dt>
<dd><p>Mapping between symbolic characters (A, B, C, X, Y, Z, _) and wavelet features.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.generator">
<span class="sig-name descname"><span class="pre">generator</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.generator" title="Link to this definition"></a></dt>
<dd><p>Name of the wavelet basis used (default: ‘ricker’).</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.compute_cwt">
<span class="sig-name descname"><span class="pre">compute_cwt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scales</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.compute_cwt" title="Link to this definition"></a></dt>
<dd><p>Computes the Continuous Wavelet Transform using the Ricker wavelet
and stores the transformed signals in <cite>transforms</cite>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.encode_dna">
<span class="sig-name descname"><span class="pre">encode_dna</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.encode_dna" title="Link to this definition"></a></dt>
<dd><p>Encodes each scale’s transformed signal into a symbolic <cite>DNAstr</cite> sequence
using local maximum coding (ABCXYZ).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.encode_dna_full">
<span class="sig-name descname"><span class="pre">encode_dna_full</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.encode_dna_full" title="Link to this definition"></a></dt>
<dd><p>Encodes signals at each scale using the full encoding scheme, preserving
the flat regions (_) and finer symbolic transitions.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.plot_codes">
<span class="sig-name descname"><span class="pre">plot_codes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scale</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.plot_codes" title="Link to this definition"></a></dt>
<dd><p>Plots both the wavelet transform and the symbolic code for the given scale.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.decode_dna">
<span class="sig-name descname"><span class="pre">decode_dna</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scale</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.decode_dna" title="Link to this definition"></a></dt>
<dd><p>Reconstructs the approximate signal for a given scale from its DNA encoding.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scale</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.__getitem__" title="Link to this definition"></a></dt>
<dd><p>Shortcut to access the <cite>DNAstr</cite> object for a specific scale.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.summary">
<span class="sig-name descname"><span class="pre">summary</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.summary" title="Link to this definition"></a></dt>
<dd><p>Returns a dictionary summarizing encoded scales and metadata.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.has">
<span class="sig-name descname"><span class="pre">has</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scale</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.has" title="Link to this definition"></a></dt>
<dd><p>Checks whether a DNAstr encoding exists at the given scale.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">Static</span> <span class="pre">pairwise</span> <span class="pre">distance</span> <span class="pre">methods</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">--------------------------------</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal._pairwiseEntropyDistance">
<span class="sig-name descname"><span class="pre">_pairwiseEntropyDistance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">list</span> <span class="pre">of</span> <span class="pre">DNAstr</span> <span class="pre">objects</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal._pairwiseEntropyDistance" title="Link to this definition"></a></dt>
<dd><p>Return a DNApairwiseAnalysis instance based on the mutually exclusive information
after DNA/code alignment.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal._pairwiseJaccardMotifDistance">
<span class="sig-name descname"><span class="pre">_pairwiseJaccardMotifDistance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">list</span> <span class="pre">of</span> <span class="pre">DNAstr</span> <span class="pre">objects</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal._pairwiseJaccardMotifDistance" title="Link to this definition"></a></dt>
<dd><p>Return a DNApairwisedistance based on the presence/absence of a pattern (default=YAZB)</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal._pairwiseJensenShannonDistance">
<span class="sig-name descname"><span class="pre">_pairwiseJensenShannonDistance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">list</span> <span class="pre">of</span> <span class="pre">DNAstr</span> <span class="pre">objects</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal._pairwiseJensenShannonDistance" title="Link to this definition"></a></dt>
<dd><p>Return a DNApairwisedistance based on the Jensen-Shannon distance of a pattern</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal._pairwiseLevenshteinDistance">
<span class="sig-name descname"><span class="pre">_pairwiseLevenshteinDistance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">list</span> <span class="pre">of</span> <span class="pre">DNAstr</span> <span class="pre">objects</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal._pairwiseLevenshteinDistance" title="Link to this definition"></a></dt>
<dd><p>Return a DNApairwisedistance based on the Levenshtein Distance</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">Methods:</span> <span class="pre">Symbolic</span> <span class="pre">Encoding</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">--------------------------</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id8">
<span class="sig-name descname"><span class="pre">encode_dna</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scales</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id8" title="Link to this definition"></a></dt>
<dd><p>Convert signal into triplet-based symbolic encoding (per scale).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id9">
<span class="sig-name descname"><span class="pre">encode_dna_full</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scales</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resolution</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'index'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">repeat</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_points</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id9" title="Link to this definition"></a></dt>
<dd><p>Generate full-resolution DNA strings by repeating letters (to <cite>codesfull</cite>).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">Methods:</span> <span class="pre">Sinusoidal</span> <span class="pre">Encoding</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">----------------------------</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.sinencode_dna">
<span class="sig-name descname"><span class="pre">sinencode_dna</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scales</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">d_part</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">32</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">N</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.sinencode_dna" title="Link to this definition"></a></dt>
<dd><p>Encode symbolic segments (from <cite>codes</cite>) into sinusoidal vectors.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.sinencode_dna_full">
<span class="sig-name descname"><span class="pre">sinencode_dna_full</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scales</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">d_part</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">32</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">N</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.sinencode_dna_full" title="Link to this definition"></a></dt>
<dd><p>Encode symbolic full strings (from <cite>codesfull</cite>) into sinusoidal vectors.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">Methods:</span> <span class="pre">Sinusoidal</span> <span class="pre">Decoding</span> <span class="pre">(Static)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">-------------------------------------</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.sindecode_dna">
<span class="sig-name descname"><span class="pre">sindecode_dna</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">grouped_embeddings</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_dx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">d_part</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">32</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">N</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.sindecode_dna" title="Link to this definition"></a></dt>
<dd><p>[static] Decode grouped sinusoidal embeddings → <cite>DNACodes</cite> structure.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.sindecode_dna_full">
<span class="sig-name descname"><span class="pre">sindecode_dna_full</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">grouped_embeddings</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_dx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">d_part</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">32</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">N</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.sindecode_dna_full" title="Link to this definition"></a></dt>
<dd><p>[static] Decode full sinusoidal embeddings → <cite>DNAFullCodes</cite> structure.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">Methods:</span> <span class="pre">Signal</span> <span class="pre">Reconstruction</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">------------------------------</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.tosignal">
<span class="sig-name descname"><span class="pre">tosignal</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">codes_attr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'codes'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.tosignal" title="Link to this definition"></a></dt>
<dd><p>Reconstruct approximate signal (as <cite>signal</cite> instance) from symbolic encoding.</p>
</dd></dl>

<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">S</span> <span class="o">=</span> <span class="n">signal</span><span class="o">.</span><span class="n">from_peaks</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>  <span class="c1"># define a signal</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dna</span> <span class="o">=</span> <span class="n">DNAsignal</span><span class="p">(</span><span class="n">S</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dna</span><span class="o">.</span><span class="n">compute_cwt</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dna</span><span class="o">.</span><span class="n">encode_dna</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dna</span><span class="o">.</span><span class="n">codes</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>
<span class="go">DNAstr(&quot;AAAZZZYY...&quot;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dna</span><span class="o">.</span><span class="n">plot_codes</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dna</span><span class="o">.</span><span class="n">codes</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;YAZB&quot;</span><span class="p">)</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.align_with">
<span class="sig-name descname"><span class="pre">align_with</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.align_with" title="Link to this definition"></a></dt>
<dd><p>Align symbolic sequences and compute mutual entropy.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><dl class="simple">
<dt>with fields</dt><dd><ul class="simple">
<li><p>seq1_aligned (str)</p></li>
<li><p>seq2_aligned (str)</p></li>
<li><p>aligned_signal (list of tuples)</p></li>
<li><p>mutual_entropy (float)</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>SimpleNamespace</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.apply_baseline_filter">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">apply_baseline_filter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">signal</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">w</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delta_t</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.apply_baseline_filter" title="Link to this definition"></a></dt>
<dd><p>Apply baseline filtering using moving median and local Poisson-based thresholding.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>signal</strong> (<em>np.ndarray</em>) – Input signal (expected to be non-negative or baseline-dominated).</p></li>
<li><p><strong>w</strong> (<em>int</em><em> or </em><em>None</em>) – Window size for baseline and statistics (must be odd).
Defaults to max(11, 2% of signal length).</p></li>
<li><p><strong>k</strong> (<em>float</em>) – Bienaymé-Tchebychev multiplier.</p></li>
<li><p><strong>delta_t</strong> (<em>float</em>) – Sampling time step.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>filtered</strong> – Signal with baseline removed and low-intensity noise suppressed.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This method is static, use signal.apply_baseline_filter() whenever appropriate instead.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id10">
<span class="sig-name descname"><span class="pre">compute_cwt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scales</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">apply_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id10" title="Link to this definition"></a></dt>
<dd><p>Compute Continuous Wavelet Transform (CWT) using the Mexican Hat wavelet.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scales</strong> (<em>list</em><em>, </em><em>int</em><em>, or </em><em>None</em>) – List of scales (or a single scale) to compute. If None, default to [1, 2, 4, 8, 16].</p></li>
<li><p><strong>apply_filter</strong> (<em>bool</em>) – Whether to apply a baseline filter to the input signal before transforming.</p></li>
<li><p><strong>Sets</strong> – </p></li>
<li><p><strong>----</strong> – </p></li>
<li><p><strong>self.scales</strong> (<em>list</em>) – The list of actual scales used.</p></li>
<li><p><strong>self.filtered_signal</strong> (<em>ndarray</em>) – Filtered or raw signal used for CWT.</p></li>
<li><p><strong>self.cwt_coeffs</strong> (<em>dict</em>) – Dictionary mapping each scale to its 1D coefficient array.</p></li>
<li><p><strong>self.transforms</strong> (<a class="reference internal" href="#sig2dna_core.signomics.signal_collection" title="sig2dna_core.signomics.signal_collection"><em>signal_collection</em></a>) – Collection of <cite>signal</cite> objects storing the transformed signals for each scale.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id11">
<span class="sig-name descname"><span class="pre">encode_dna</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scales</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id11" title="Link to this definition"></a></dt>
<dd><p>Encode each transformed signal into a symbolic DNA-like sequence of monotonic segments.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scales</strong> (<em>list</em><em>, </em><em>int</em><em>, or </em><em>None</em>) – List of scales (or a single scale) to encode. If None, use self.scales.</p></li>
<li><p><strong>letters</strong> (<em>The encoding detects strictly monotonic</em><em> (or </em><em>flat</em><em>) </em><em>segments and labels them with symbolic</em>) – </p></li>
<li><p><strong>A</strong> (<em>-</em>) – </p></li>
<li><p><strong>Z</strong> (<em>-</em>) – </p></li>
<li><p><strong>B</strong> (<em>-</em>) – </p></li>
<li><p><strong>Y</strong> (<em>-</em>) – </p></li>
<li><p><strong>C</strong> (<em>-</em>) – </p></li>
<li><p><strong>X</strong> (<em>-</em>) – </p></li>
<li><p><strong>_</strong> (<em>-</em>) – </p></li>
<li><p><strong>Sets</strong> – </p></li>
<li><p><strong>----</strong> – </p></li>
<li><p><strong>self.codes</strong> (<em>dict</em>) – <dl class="simple">
<dt>Dictionary mapping each scale to a struct with:</dt><dd><ul>
<li><p>letters : str (symbolic encoding)</p></li>
<li><p>widths  : list of float (x-span of each segment)</p></li>
<li><p>heights : list of float (y-delta of each segment)</p></li>
<li><p>iloc    : list of index-pair tuples (start, end+1)</p></li>
<li><p>xloc    : list of x-span tuples (x_start, x_end)</p></li>
<li><p>dx      : segment step (dx)</p></li>
</ul>
</dd>
</dl>
</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id12">
<span class="sig-name descname"><span class="pre">encode_dna_full</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scales</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resolution</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'index'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">repeat</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_points</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id12" title="Link to this definition"></a></dt>
<dd><p>Convert symbolic codes into DNA-like strings by repeating letters proportionally to their span.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scales</strong> (<em>list</em><em>, </em><em>int</em><em>, or </em><em>None</em>) – List of scales (or a single scale) to convert. If None, use self.scales.</p></li>
<li><p><strong>resolution</strong> (<em>{'index'</em><em>, </em><em>'x'}</em>) – <dl class="simple">
<dt>Repetition mode:</dt><dd><ul>
<li><p>’index’: repeat letters by number of indices (j - i from iloc)</p></li>
<li><p>’x’    : interpolate letter values over physical x-axis distance (xloc)</p></li>
</ul>
</dd>
</dl>
</p></li>
<li><p><strong>repeat</strong> (<em>bool</em>) – If True, repeat or interpolate letters to form a string of desired resolution.
If False, return the symbolic sequence without repetition.</p></li>
<li><p><strong>n_points</strong> (<em>int</em><em> or </em><em>None</em>) – Used only for resolution=’x’ to control the number of interpolation points.
If None, defaults to ~10 points per x-unit.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>dict</em> – Dictionary mapping each scale to its DNA-like string.</p></li>
<li><p><em>Sets</em></p></li>
<li><p><em>—-</em></p></li>
<li><p><strong>self.codesfull</strong> (<em>dict</em>) – Dictionary storing the resulting full DNA-like string per scale.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.entropy_from_string">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">entropy_from_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.entropy_from_string" title="Link to this definition"></a></dt>
<dd><p>return the entropy of a string</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.find_sequence">
<span class="sig-name descname"><span class="pre">find_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pattern</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.find_sequence" title="Link to this definition"></a></dt>
<dd><p>Find occurrences of a specific letter pattern in encoded sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.get_code">
<span class="sig-name descname"><span class="pre">get_code</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scale</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.get_code" title="Link to this definition"></a></dt>
<dd><p>Retrieve encoded data for a specific scale.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.get_entropy">
<span class="sig-name descname"><span class="pre">get_entropy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scale</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.get_entropy" title="Link to this definition"></a></dt>
<dd><p>Calculate Shannon entropy for encoded signal.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id13">
<span class="sig-name descname"><span class="pre">has</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scale</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id13" title="Link to this definition"></a></dt>
<dd><p>Check if a DNA encoding exists for the specified scale.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>scale</strong> (<em>int</em>) – The wavelet scale to check.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if a symbolic DNAstr encoding exists at the given scale, False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">dna</span><span class="o">.</span><span class="n">has</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dna</span><span class="o">.</span><span class="n">has</span><span class="p">(</span><span class="mi">16</span><span class="p">)</span>
<span class="go">False</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id14">
<span class="sig-name descname"><span class="pre">plot_codes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scale</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colormap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.4</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id14" title="Link to this definition"></a></dt>
<dd><p>Plot the symbolic DNA-like encoding as colored triangle segments.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scale</strong> (<em>int</em>) – The scale at which the signal was encoded.</p></li>
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em><em>, </em><em>optional</em>) – Axis to draw on. If None, a new figure is created.</p></li>
<li><p><strong>colormap</strong> (<em>dict</em><em>, </em><em>optional</em>) – Custom mapping of letters to colors. Default uses 7 distinct colors.</p></li>
<li><p><strong>alpha</strong> (<em>float</em>) – Transparency for the patches. Default is 0.4.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.plot_signals">
<span class="sig-name descname"><span class="pre">plot_signals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scales</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.plot_signals" title="Link to this definition"></a></dt>
<dd><p>Plot signals.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.plot_transforms">
<span class="sig-name descname"><span class="pre">plot_transforms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">indices</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.plot_transforms" title="Link to this definition"></a></dt>
<dd><p>Plot the stored CWT-transformed signals as a signal collection.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>indices</strong> (<em>list</em><em>[</em><em>int</em><em> or </em><em>str</em><em>]</em><em>, </em><em>optional</em>) – Specific scales or names to plot.</p></li>
<li><p><strong>kwargs</strong> (passed to <cite>signal_collection.plot</cite>) – </p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.print_alignment">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">print_alignment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">80</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.print_alignment" title="Link to this definition"></a></dt>
<dd><p>print aligned sequences</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.reconstruct_aligned_string">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">reconstruct_aligned_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aligned</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.reconstruct_aligned_string" title="Link to this definition"></a></dt>
<dd><p>Fast reconstruction of aligned signals</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.reconstruct_signal">
<span class="sig-name descname"><span class="pre">reconstruct_signal</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scale</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_signal</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.reconstruct_signal" title="Link to this definition"></a></dt>
<dd><p>Reconstruct the signal from symbolic features (e.g., YAZB).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scale</strong> (<em>int</em>) – Scale to use for reconstruction.</p></li>
<li><p><strong>return_signal</strong> (<em>bool</em>) – If True, return a <cite>signal</cite> object. Else return y array.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Reconstructed signal.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sig2dna_core.signomics.signal" title="sig2dna_core.signomics.signal">signal</a> or np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id15">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sindecode_dna</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">grouped_embeddings</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_dx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">d_part</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">32</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">N</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id15" title="Link to this definition"></a></dt>
<dd><p>Decode sinusoidal grouped embeddings into a DNACodes structure.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>grouped_embeddings</strong> (<em>dict</em>) – {scale: {letter: np.ndarray}} sinusoidal representations</p></li>
<li><p><strong>reference_dx</strong> (<em>float</em>) – Sampling resolution used to reconstruct xloc and iloc</p></li>
<li><p><strong>d_part</strong> (<em>int</em>) – Dimensionality per component (start, width, height)</p></li>
<li><p><strong>N</strong> (<em>int</em>) – Frequency base</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Decoded symbolic structure</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sig2dna_core.signomics.DNACodes" title="sig2dna_core.signomics.DNACodes">DNACodes</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id16">
<span class="sig-name descname"><span class="pre">sinencode_dna</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scales</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">d_part</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">32</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">N</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id16" title="Link to this definition"></a></dt>
<dd><p>Encode <cite>self.codes</cite> into sinusoidal embeddings (grouped by letter).</p>
<dl class="simple">
<dt>Sets:</dt><dd><ul class="simple">
<li><p>self.codes (DNACodes): Encoded version of original codes.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id17">
<span class="sig-name descname"><span class="pre">sinencode_dna_full</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d_model</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">96</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">N</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id17" title="Link to this definition"></a></dt>
<dd><p>🌀 Encode full-resolution DNA-like strings into sinusoidal embeddings grouped by letter.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>d_model</strong> (<em>int</em>) – Dimensionality of the sinusoidal embedding (must be even).</p></li>
<li><p><strong>N</strong> (<em>int</em>) – Frequency base used for sinusoidal encoding.</p></li>
<li><p><strong>Sets</strong> – </p></li>
<li><p><strong>----</strong> – </p></li>
<li><p><strong>self.codesfull</strong> (<a class="reference internal" href="#sig2dna_core.signomics.DNAFullCodes" title="sig2dna_core.signomics.DNAFullCodes"><em>DNAFullCodes</em></a><em> (</em><em>if not already set</em><em>)</em>) – Full-resolution symbolic strings at each scale.</p></li>
<li><p><strong>self.codesfull_encoded</strong> (<a class="reference internal" href="#sig2dna_core.signomics.DNAFullCodes" title="sig2dna_core.signomics.DNAFullCodes"><em>DNAFullCodes</em></a>) – Sinusoidally encoded version of the full DNA strings.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAsignal.synthetic_signal">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">synthetic_signal</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">peaks</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">baseline</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAsignal.synthetic_signal" title="Link to this definition"></a></dt>
<dd><p>Generate flexible synthetic signals. (obsolete)</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id18">
<span class="sig-name descname"><span class="pre">tosignal</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">codes_attr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'codes'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id18" title="Link to this definition"></a></dt>
<dd><p>Reconstruct an approximate signal from symbolic encodings.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scale</strong> (<em>int</em><em> or </em><em>None</em>) – Scale level to use (defaults to first available if None).</p></li>
<li><p><strong>codes_attr</strong> (<em>str</em>) – Attribute from which to decode (‘codes’ or ‘codesfull’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An approximate <cite>signal</cite> object reconstructed from symbolic information.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sig2dna_core.signomics.signal" title="sig2dna_core.signomics.signal">signal</a></p>
</dd>
</dl>
</dd></dl>

</section>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sig2dna_core.signomics.</span></span><span class="sig-name descname"><span class="pre">DNAstr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">content</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iloc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xloc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'index'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_unit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'-'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">engine</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'difflib'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">engineOpts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></p>
<p>A symbolic DNA-like sequence class supporting alignment, entropy analysis,
edit-distance metrics, and signal reconstruction from symbolic codes.</p>
<p>Extended from <cite>str</cite>, it is designed for symbolic transformations of signals
(e.g., wavelet-encoded GC-MS peaks or time series).</p>
<section id="main-features">
<h2>Main Features<a class="headerlink" href="#main-features" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>Supports symbolic operations for pattern recognition, entropy, alignment.</p></li>
<li><p>Encodes x-resolution (<cite>dx</cite>), original index (<cite>iloc</cite>), and physical x-range (<cite>xloc</cite>).</p></li>
<li><p>Aligns sequences with visual inspection and rich diffs.</p></li>
<li><p>Converts symbolic strings into synthetic numerical signals.</p></li>
</ul>
</section>
<section id="operators">
<h2>Operators<a class="headerlink" href="#operators" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>: concatenate two DNAstr objects</p></li>
</ul>
<ul class="simple">
<li><p>: symbolic difference after alignment (mismatches only)</p></li>
</ul>
<p>== : equality comparison (exact content and dx)</p>
</section>
<section id="key-methods">
<h2>Key Methods<a class="headerlink" href="#key-methods" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>__init__ / __new__      : Constructor with metadata (<cite>dx</cite>, <cite>iloc</cite>, <cite>xloc</cite>)</p></li>
<li><p>align(other)            : Align this DNAstr to another, update mask and aligned views</p></li>
<li><p>wrapped_alignment()     : Pretty terminal view of the alignment with colors and symbols</p></li>
<li><p>html_alignment()        : Rich HTML display of the alignment (Jupyter)</p></li>
<li><p>plot_alignment()        : Visualize waveform alignment with symbolic signals</p></li>
<li><p>plot_mask()             : Color block plot showing matches/mismatches/gaps</p></li>
<li><p>find(pattern, regex=False) : Search for symbolic patterns with fuzziness or regex</p></li>
<li><p>to_signal()             : Convert symbolic code into synthetic signal (NumPy)</p></li>
<li><p>vectorized()            : Convert string to integer codes</p></li>
<li><p>summary()               : Print entropy and character frequencies</p></li>
<li><p>mutation_counts         : Property: {‘matches’, ‘mismatches’, ‘indels’}</p></li>
<li><p>entropy                 : Property: Shannon entropy</p></li>
<li><p>mutual_entropy(other)   : Mutual entropy of two sequences</p></li>
<li><p>excess_entropy(other)   : Excess entropy H1 + H2 - 2 * H12</p></li>
<li><p>jensen_shannon(other)   : Jensen-Shannon divergence</p></li>
<li><p>jaccard(other)          : Jaccard similarity</p></li>
<li><p>alignment_stats         : Property: Match, substitution, gap counts</p></li>
<li><p>score(normalized=True) : Alignment score (fraction of matches)</p></li>
<li><p>has(other: str)         : Check if a pattern or substring exists</p></li>
</ul>
<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.dx">
<span class="sig-name descname"><span class="pre">dx</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.dx" title="Link to this definition"></a></dt>
<dd><p>Average resolution along the x-axis.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.iloc">
<span class="sig-name descname"><span class="pre">iloc</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.iloc" title="Link to this definition"></a></dt>
<dd><p>Positional index or index range in the source DNA string.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int or tuple of int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.xloc">
<span class="sig-name descname"><span class="pre">xloc</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.xloc" title="Link to this definition"></a></dt>
<dd><p>Corresponding x-value(s) for the symbolic sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float or tuple of float</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.aligned_with">
<span class="sig-name descname"><span class="pre">aligned_with</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.aligned_with" title="Link to this definition"></a></dt>
<dd><p>Aligned form of self with insertions (spaces) where needed.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.other_copy">
<span class="sig-name descname"><span class="pre">other_copy</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.other_copy" title="Link to this definition"></a></dt>
<dd><p>Aligned form of the reference sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.ref_hash">
<span class="sig-name descname"><span class="pre">ref_hash</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.ref_hash" title="Link to this definition"></a></dt>
<dd><p>SHA256 hash of the aligned reference sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.mask">
<span class="sig-name descname"><span class="pre">mask</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.mask" title="Link to this definition"></a></dt>
<dd><p>Alignment mask: ‘=’ for matches, ‘*’ for substitutions, ‘ ‘ for gaps.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.engine">
<span class="sig-name descname"><span class="pre">engine</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.engine" title="Link to this definition"></a></dt>
<dd><p>Alignment engine: ‘difflib’ or ‘bio’.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.engineOpts">
<span class="sig-name descname"><span class="pre">engineOpts</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.engineOpts" title="Link to this definition"></a></dt>
<dd><p>Options passed to the alignment engine.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s1</span> <span class="o">=</span> <span class="n">DNAstr</span><span class="p">(</span><span class="s2">&quot;YYAAZZBB&quot;</span><span class="p">,</span> <span class="n">dx</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s2</span> <span class="o">=</span> <span class="n">DNAstr</span><span class="p">(</span><span class="s2">&quot;YAABZBB&quot;</span><span class="p">,</span> <span class="n">dx</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s1</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">s2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">s1</span><span class="o">.</span><span class="n">wrapped_alignment</span><span class="p">(</span><span class="mi">40</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s1</span><span class="o">.</span><span class="n">plot_alignment</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">segments</span> <span class="o">=</span> <span class="n">s1</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;YAZB&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">segments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">to_signal</span><span class="p">()</span><span class="o">.</span><span class="n">plot</span><span class="p">()</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.align">
<span class="sig-name descname"><span class="pre">align</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">engine</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">engineOpts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">forced</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.align" title="Link to this definition"></a></dt>
<dd><p>Align this DNAstr sequence to another, allowing insertions/deletions to maximize matches.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="#sig2dna_core.signomics.DNAstr" title="sig2dna_core.signomics.DNAstr"><em>DNAstr</em></a>) – Another DNAstr object to align with.</p></li>
<li><p><strong>engine</strong> (<em>{'difflib'</em><em>, </em><em>'bio'}</em><em> or </em><em>None</em>) – <dl class="simple">
<dt>Alignment engine to use:</dt><dd><ul>
<li><p>’difflib’: uses difflib.SequenceMatcher (fast, approximate).</p></li>
<li><p>’bio’   : uses Bio.Align.PairwiseAligner (biologically inspired global alignment).</p></li>
</ul>
</dd>
</dl>
<p>If None, defaults to self.engine.</p>
</p></li>
<li><p><strong>engineOpts</strong> (<em>dict</em><em>, </em><em>optional</em>) – Dictionary of alignment parameters for the selected engine.</p></li>
<li><p><strong>forced</strong> (<em>bool</em>) – If True, allow alignment even if <cite>dx</cite> values differ. If False (default), a mismatch in
<cite>dx</cite> will raise an error to prevent incorrect alignment of signals with different sampling.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>aligned_self</strong> (<em>str</em>) – Aligned version of this sequence (with gaps inserted where needed).</p></li>
<li><p><strong>aligned_other</strong> (<em>str</em>) – Aligned version of the other sequence.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The alignment is symmetric and permanent: both sequences are aligned with
gaps introduced (spaces) to preserve positional correspondence. A hash of
the aligned <cite>other</cite> sequence is stored to detect redundant alignments.</p>
<dl class="simple">
<dt>A match mask (<cite>self.mask</cite>) is generated with:</dt><dd><p>‘=’ for exact matches,
‘*’ for mismatches (substitutions),
‘ ‘ for insertions/deletions (gaps).</p>
</dd>
<dt>The method updates:</dt><dd><ul class="simple">
<li><p>self.aligned_with</p></li>
<li><p>self.other_copy</p></li>
<li><p>self.mask</p></li>
<li><p>self.ref_hash</p></li>
</ul>
</dd>
</dl>
<section id="example">
<h3>Example:<a class="headerlink" href="#example" title="Link to this heading"></a></h3>
<p>S1 = DNAstr(“AABBCC”)
S2 = DNAstr(“AACBCC”)
S1.align(S2,”difflib”)
print(S1.mask)
print(S1.wrapped_alignment())
==*===
AACBCC
|| |||
AABBCC</p>
<p>S1 = DNAstr(“AABBCC”)
S2 = DNAstr(“AACBCC”)
S1.align(S2,”bio”)
print(S1.mask)
print(S1.wrapped_alignment())
==  ==
AAB·CC
||  ||
AA·BCC</p>
<p>S1 = DNAstr(“AABBCCXYZZZ”)
S2 = DNAstr(“AACBCCZZXXX”)
S1.align(S2,”bio”)
print(S1.mask)
print(S1.wrapped_alignment())
== *   ==
AABCC··ZZ
||     ||
AA·B·CCZZ</p>
</section>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.aligned_code">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">aligned_code</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.aligned_code" title="Link to this definition"></a></dt>
<dd><p>return aligned code</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.alignment_stats">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">alignment_stats</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.alignment_stats" title="Link to this definition"></a></dt>
<dd><p>Retrun DNAstr alignment statistics</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.entropy">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">entropy</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.entropy" title="Link to this definition"></a></dt>
<dd><p>Compute the Shannon entropy of the DNAstr sequence</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.excess_entropy">
<span class="sig-name descname"><span class="pre">excess_entropy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.excess_entropy" title="Link to this definition"></a></dt>
<dd><p>Compute the excess Shannon entropy of two DNAstr sequences H(A)+H(B)-2*H(AB)</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.extract_motifs">
<span class="sig-name descname"><span class="pre">extract_motifs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pattern</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'YAZB'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">minlen</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.extract_motifs" title="Link to this definition"></a></dt>
<dd><p>Extract and analyze YAZB motifs (canonical and distorted) from the symbolic sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pattern</strong> (<em>str</em>) – Canonical motif pattern (default is ‘YAZB’).</p></li>
<li><p><strong>minlen</strong> (<em>int</em>) – Minimum motif length to be considered valid.</p></li>
<li><p><strong>plot</strong> (<em>bool</em>) – If True, generate a motif density plot using xloc or sequence index.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Table of detected motifs with start/end positions, length, and classification.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.find">
<span class="sig-name descname"><span class="pre">find</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pattern</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regex</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.find" title="Link to this definition"></a></dt>
<dd><p>Finds all fuzzy (or regex-based) occurrences of a DNA-like sequence pattern.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pattern</strong> (<em>str</em>) – The symbolic sequence to search for (e.g., “YAZB”).</p></li>
<li><p><strong>regex</strong> (<em>bool</em><em>, </em><em>optional</em>) – If False (default), interprets pattern as symbolic and inserts ‘.’ between characters.
If True, uses the raw pattern as a regular expression.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>A list of DNAstr slices with attributes:</dt><dd><ul class="simple">
<li><p>iloc: (start_idx, end_idx)</p></li>
<li><p>xloc: (x_start, x_end)</p></li>
<li><p>width: segment width</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list of <a class="reference internal" href="#sig2dna_core.signomics.DNAstr" title="sig2dna_core.signomics.DNAstr">DNAstr</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.html_alignment">
<span class="sig-name descname"><span class="pre">html_alignment</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.html_alignment" title="Link to this definition"></a></dt>
<dd><p>Render the alignment using HTML with color coding:
- green: match
- blue: gap
- red: substitution</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Displays HTML directly in Jupyter/Notebook environments.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.jaccard">
<span class="sig-name descname"><span class="pre">jaccard</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.jaccard" title="Link to this definition"></a></dt>
<dd><p>Compute the Jaccard distance between two DNAstr sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>other</strong> (<a class="reference internal" href="#sig2dna_core.signomics.DNAstr" title="sig2dna_core.signomics.DNAstr"><em>DNAstr</em></a>) – The other DNAstr sequence to compare with.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Jaccard distance: 1 - (intersection / union) of unique letters.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.jensen_shannon">
<span class="sig-name descname"><span class="pre">jensen_shannon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.jensen_shannon" title="Link to this definition"></a></dt>
<dd><p>Compute the Jensen-Shannon distance between self and another DNAstr.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="#sig2dna_core.signomics.DNAstr" title="sig2dna_core.signomics.DNAstr"><em>DNAstr</em></a>) – Another DNAstr instance.</p></li>
<li><p><strong>base</strong> (<em>float</em><em>, </em><em>optional</em>) – Base for the logarithm (default: 2)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Jensen-Shannon distance.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.levenshtein">
<span class="sig-name descname"><span class="pre">levenshtein</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_alignment</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">engine</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">engineOpts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">forced</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.levenshtein" title="Link to this definition"></a></dt>
<dd><p>Compute the Levenshtein distance between this DNAstr and another one.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="#sig2dna_core.signomics.DNAstr" title="sig2dna_core.signomics.DNAstr"><em>DNAstr</em></a>) – Another DNAstr object to compare against.</p></li>
<li><p><strong>use_alignment</strong> (<em>bool</em><em>, </em><em>default=True</em>) – If True, uses the aligned sequences (computed if necessary).
If False, compares the raw sequences directly.</p></li>
<li><p><strong>engine</strong> (<em>{'difflib'</em><em>, </em><em>'bio'}</em><em>, </em><em>optional</em>) – Alignment engine to use if alignment is needed.</p></li>
<li><p><strong>engineOpts</strong> (<em>dict</em><em>, </em><em>optional</em>) – Parameters for the selected alignment engine.</p></li>
<li><p><strong>forced</strong> (<em>bool</em><em>, </em><em>default=False</em>) – Force alignment even if dx values differ.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>dist</strong> – Levenshtein distance between the two sequences (aligned or raw).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>A = DNAstr(“YAZBZAY”)
B = DNAstr(“YAZBZZY”)
A.levenshtein_distance(B, use_alignment=False)  # raw
A.levenshtein_distance(B, use_alignment=True, engine=”bio”)  # aligned</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.mutation_counts">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mutation_counts</span></span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.mutation_counts" title="Link to this definition"></a></dt>
<dd><p>Counts of insertions, deletions/substitutions, and matches.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.mutual_entropy">
<span class="sig-name descname"><span class="pre">mutual_entropy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.mutual_entropy" title="Link to this definition"></a></dt>
<dd><p>Compute the Shannon mutual entropy of two DNAstr sequences from their aligned segments</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.plot_alignment">
<span class="sig-name descname"><span class="pre">plot_alignment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.plot_alignment" title="Link to this definition"></a></dt>
<dd><p>Plot a block alignment view of two DNAstr sequences with color-coded segments.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dx</strong> (<em>float</em>) – Horizontal step between segments (defaults to 1.0).</p></li>
<li><p><strong>dy</strong> (<em>float</em>) – Vertical height increment for symbolic waveform visualization.</p></li>
<li><p><strong>width</strong> (<em>int</em>) – Number of characters per row (line wrapping).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>matplotlib.figure.Figure</em></p></li>
<li><p><em>matplotlib.axes.Axes</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.plot_mask">
<span class="sig-name descname"><span class="pre">plot_mask</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.plot_mask" title="Link to this definition"></a></dt>
<dd><p>Plot a color-coded mask of the alignment between sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Matplotlib figure of the alignment mask.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>matplotlib.figure.Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.score">
<span class="sig-name descname"><span class="pre">score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">normalized</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.score" title="Link to this definition"></a></dt>
<dd><p>Return an alignment score, optionally normalized.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>normalized</strong> (<em>bool</em>) – If True (default), return score as a fraction of total aligned positions.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Alignment score.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.summary">
<span class="sig-name descname"><span class="pre">summary</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.summary" title="Link to this definition"></a></dt>
<dd><p>Summarize the DNAstr with key stats: length, unique letters, entropy, etc.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary containing length, letter frequency, Shannon entropy, and dx.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.to_signal">
<span class="sig-name descname"><span class="pre">to_signal</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.to_signal" title="Link to this definition"></a></dt>
<dd><p>Converts the symbolic DNA sequence into a synthetic NumPy array mimicking the original wavelet-transformed signal.</p>
<dl class="simple">
<dt>Rules per letter:</dt><dd><ul class="simple">
<li><p>‘A’: Crosses zero upward → linear from -1 to +1, zero in the middle</p></li>
<li><p>‘Z’: Crosses zero downward → linear from +1 to -1, zero in the middle</p></li>
<li><p>‘B’: Increasing negative → from -1 to 0</p></li>
<li><p>‘Y’: Decreasing negative → from 0 to -1</p></li>
<li><p>‘C’: Increasing positive → from 0 to +1</p></li>
<li><p>‘X’: Decreasing positive → from +1 to 0</p></li>
<li><p>‘_’: Flat at 0</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Synthetic signal array matching the symbolic encoding.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.vectorized">
<span class="sig-name descname"><span class="pre">vectorized</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">codebook</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'A':</span> <span class="pre">1,</span> <span class="pre">'B':</span> <span class="pre">2,</span> <span class="pre">'C':</span> <span class="pre">3,</span> <span class="pre">'X':</span> <span class="pre">4,</span> <span class="pre">'Y':</span> <span class="pre">5,</span> <span class="pre">'Z':</span> <span class="pre">6,</span> <span class="pre">'_':</span> <span class="pre">0}</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.vectorized" title="Link to this definition"></a></dt>
<dd><p>Map the DNAstr content to an integer array using a codebook.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>codebook</strong> (<em>dict</em><em>, </em><em>optional</em>) – Dictionary mapping characters to integer values.
default = {“A”:1,”B”:2,”C”:3,”X”:4,”Y”:5,”Z”:6,”_”:0}
None will generate a codebook based on current symbols only</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Vectorized integer representation of the string.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.DNAstr.wrapped_alignment">
<span class="sig-name descname"><span class="pre">wrapped_alignment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">80</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.DNAstr.wrapped_alignment" title="Link to this definition"></a></dt>
<dd><p>Return a line-wrapped alignment view (multi-line), optionally color-coded
for terminal/IPython usage (Spyder, Jupyter).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>width</strong> (<em>int</em>) – Number of characters per line in wrapped display.</p></li>
<li><p><strong>colors</strong> (<em>bool</em>) – If True, use ANSI codes to highlight differences. May be overridden
if terminal does not support ANSI (e.g., Spyder).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Wrapped, optionally colorized alignment.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

</section>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sig2dna_core.signomics.SinusoidalEncoder">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sig2dna_core.signomics.</span></span><span class="sig-name descname"><span class="pre">SinusoidalEncoder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d_model=96</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">N=10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype=&lt;class</span> <span class="pre">'numpy.float32'&gt;</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.SinusoidalEncoder" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>🌀 Generic sinusoidal encoder/decoder supporting symbolic and numeric sequences.</p>
<p>Each scalar value is transformed into a vector of dimension <cite>d_model</cite>, where alternating components
contain sinusoidal features of increasing frequency. The mapping is based on:</p>
<blockquote>
<div><dl class="simple">
<dt>For k = 0 to d_model/2 - 1:</dt><dd><p>f_{2k}(x)   = sin(x / r_k)
f_{2k+1}(x) = cos(x / r_k)</p>
</dd>
</dl>
<p>where r_k = N^(2k / d_model)</p>
</div></blockquote>
<p>This representation preserves relative positions and scaling in a smooth, topologically faithful
embedding space. The class supports multiple decoding strategies, scaling logic, residual control,
and round-trip verification.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>d_model</strong> (<em>int</em>) – Dimensionality of each sinusoidal encoding (must be even).</p></li>
<li><p><strong>N</strong> (<em>int</em>) – Frequency base for the positional encoding.</p></li>
<li><p><strong>dtype</strong> (<em>np.dtype</em>) – Output data type (default: np.float32).</p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.SinusoidalEncoder.d_model">
<span class="sig-name descname"><span class="pre">d_model</span></span><a class="headerlink" href="#sig2dna_core.signomics.SinusoidalEncoder.d_model" title="Link to this definition"></a></dt>
<dd><p>Embedding dimensionality.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.SinusoidalEncoder.N">
<span class="sig-name descname"><span class="pre">N</span></span><a class="headerlink" href="#sig2dna_core.signomics.SinusoidalEncoder.N" title="Link to this definition"></a></dt>
<dd><p>Frequency base.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.SinusoidalEncoder.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#sig2dna_core.signomics.SinusoidalEncoder.dtype" title="Link to this definition"></a></dt>
<dd><p>Data type for encoded output.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.dtype</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.SinusoidalEncoder._last_input_type">
<span class="sig-name descname"><span class="pre">_last_input_type</span></span><a class="headerlink" href="#sig2dna_core.signomics.SinusoidalEncoder._last_input_type" title="Link to this definition"></a></dt>
<dd><p>Last input type passed to <cite>encode</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>type</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.SinusoidalEncoder._last_input_length">
<span class="sig-name descname"><span class="pre">_last_input_length</span></span><a class="headerlink" href="#sig2dna_core.signomics.SinusoidalEncoder._last_input_length" title="Link to this definition"></a></dt>
<dd><p>Last input length passed to <cite>encode</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.SinusoidalEncoder._scale">
<span class="sig-name descname"><span class="pre">_scale</span></span><a class="headerlink" href="#sig2dna_core.signomics.SinusoidalEncoder._scale" title="Link to this definition"></a></dt>
<dd><p>Scaling factor applied to normalize input values.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.SinusoidalEncoder._auto_scale_enabled">
<span class="sig-name descname"><span class="pre">_auto_scale_enabled</span></span><a class="headerlink" href="#sig2dna_core.signomics.SinusoidalEncoder._auto_scale_enabled" title="Link to this definition"></a></dt>
<dd><p>Whether autoscaling is enabled.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.SinusoidalEncoder._decode_residual_tolerance">
<span class="sig-name descname"><span class="pre">_decode_residual_tolerance</span></span><a class="headerlink" href="#sig2dna_core.signomics.SinusoidalEncoder._decode_residual_tolerance" title="Link to this definition"></a></dt>
<dd><p>Tolerance for residual error checking in decode verification.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.SinusoidalEncoder.encode">
<span class="sig-name descname"><span class="pre">encode</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.SinusoidalEncoder.encode" title="Link to this definition"></a></dt>
<dd><p>Encodes a sequence of values (scalar or symbolic) into sinusoidal embeddings.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.SinusoidalEncoder.decode">
<span class="sig-name descname"><span class="pre">decode</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">embedding</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'least_squares'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_error</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.SinusoidalEncoder.decode" title="Link to this definition"></a></dt>
<dd><p>Decodes the embedding to the original values using the selected inverse method.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.SinusoidalEncoder.fit_encoder">
<span class="sig-name descname"><span class="pre">fit_encoder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10.0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.SinusoidalEncoder.fit_encoder" title="Link to this definition"></a></dt>
<dd><p>Automatically estimates and stores a scaling factor to normalize input values.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.SinusoidalEncoder.set_decode_tolerance">
<span class="sig-name descname"><span class="pre">set_decode_tolerance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tol</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.SinusoidalEncoder.set_decode_tolerance" title="Link to this definition"></a></dt>
<dd><p>Sets the residual error threshold above which decoding results will raise a warning.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.SinusoidalEncoder.verify_roundtrip">
<span class="sig-name descname"><span class="pre">verify_roundtrip</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'least_squares'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'auto'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_details</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.SinusoidalEncoder.verify_roundtrip" title="Link to this definition"></a></dt>
<dd><p>Checks round-trip accuracy of encoding and decoding. Warns if residuals exceed tolerance.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.SinusoidalEncoder.sinencode_dna_grouped">
<span class="sig-name descname"><span class="pre">sinencode_dna_grouped</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">code</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">d_part</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">N</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.SinusoidalEncoder.sinencode_dna_grouped" title="Link to this definition"></a></dt>
<dd><p>Encodes a <cite>codes</cite> entry (triplet-based segments grouped by letter).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.SinusoidalEncoder.sindecode_dna_grouped">
<span class="sig-name descname"><span class="pre">sindecode_dna_grouped</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">grouped</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_code</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">d_part</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">N</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.SinusoidalEncoder.sindecode_dna_grouped" title="Link to this definition"></a></dt>
<dd><p>Reconstructs code dictionary from grouped sinusoidal embeddings.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.SinusoidalEncoder.sinencode_dnafull_grouped">
<span class="sig-name descname"><span class="pre">sinencode_dnafull_grouped</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dnafull</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">d_model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">N</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.SinusoidalEncoder.sinencode_dnafull_grouped" title="Link to this definition"></a></dt>
<dd><p>Encodes <cite>codesfull</cite> entries (strings) into grouped embeddings.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.SinusoidalEncoder.sindecode_dnafull_grouped">
<span class="sig-name descname"><span class="pre">sindecode_dnafull_grouped</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">grouped</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.SinusoidalEncoder.sindecode_dnafull_grouped" title="Link to this definition"></a></dt>
<dd><p>Decodes grouped full embeddings back to <cite>DNAstr</cite>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">Static</span> <span class="pre">Methods</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">--------------</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.SinusoidalEncoder.to_complex">
<span class="sig-name descname"><span class="pre">to_complex</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">emb</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.SinusoidalEncoder.to_complex" title="Link to this definition"></a></dt>
<dd><p>Convert a sinusoidal embedding (sin, cos) into complex numbers using Euler’s identity.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.SinusoidalEncoder.complex_distance">
<span class="sig-name descname"><span class="pre">complex_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">emb1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">emb2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'L2'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.SinusoidalEncoder.complex_distance" title="Link to this definition"></a></dt>
<dd><p>Compute pointwise distances between two embeddings in the complex sinusoidal space.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.SinusoidalEncoder.angle_difference">
<span class="sig-name descname"><span class="pre">angle_difference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">emb</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.SinusoidalEncoder.angle_difference" title="Link to this definition"></a></dt>
<dd><p>Compute angular differences Δθ between consecutive elements of a sinusoidal embedding.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.SinusoidalEncoder.phase_alignment">
<span class="sig-name descname"><span class="pre">phase_alignment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">emb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.SinusoidalEncoder.phase_alignment" title="Link to this definition"></a></dt>
<dd><p>Align the phase of an embedding <cite>emb</cite> to a reference embedding <cite>ref</cite> using complex phase factors.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.SinusoidalEncoder.pairwise_similarity">
<span class="sig-name descname"><span class="pre">pairwise_similarity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">emb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metric</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cosine'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.SinusoidalEncoder.pairwise_similarity" title="Link to this definition"></a></dt>
<dd><p>Compute a pairwise similarity or distance matrix (‘cosine’ or ‘L2’) between all elements.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.SinusoidalEncoder.group_centroid">
<span class="sig-name descname"><span class="pre">group_centroid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">emb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_std</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.SinusoidalEncoder.group_centroid" title="Link to this definition"></a></dt>
<dd><p>Compute the centroid (and optionally standard deviation) of groups in complex embedding space.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.SinusoidalEncoder.phase_unwrap">
<span class="sig-name descname"><span class="pre">phase_unwrap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">emb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.SinusoidalEncoder.phase_unwrap" title="Link to this definition"></a></dt>
<dd><p>Perform phase unwrapping (à la Fourier) on the sinusoidal embedding, optionally normalized to [0, 1].</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.SinusoidalEncoder.Example">
<span class="sig-name descname"><span class="pre">Example</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">without</span> <span class="pre">scaling</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.SinusoidalEncoder.Example" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">--------------------------</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">s</span> <span class="pre">=</span> <span class="pre">SinusoidalEncoder(8,</span> <span class="pre">100)</span>&#160; <span class="pre">#</span> <span class="pre">poor</span> <span class="pre">encoder</span> <span class="pre">(8</span> <span class="pre">dimensions,</span> <span class="pre">high</span> <span class="pre">N)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">a</span> <span class="pre">=</span> <span class="pre">s.encode([0,</span> <span class="pre">1,</span> <span class="pre">1,</span> <span class="pre">2,</span> <span class="pre">2,</span> <span class="pre">3,</span> <span class="pre">4,</span> <span class="pre">5,</span> <span class="pre">6,</span> <span class="pre">6])</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">s.decode(a)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">Output:</span></span></dt>
<dd><dl class="simple">
<dt>[0.0,</dt><dd><p>1.0000000072927564,
1.0000000072927564,
1.9999999989612762,
1.9999999989612762,
2.9999999881593866,
3.9999997833243714,
5.000000005427378,
5.999999552487366,
5.999999552487366]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">Notes:</span></span></dt>
<dd><ul class="simple">
<li><p>Use lower N (e.g., N = 1000) to compress phase variation and allow larger input range.</p></li>
<li><p>Use scaling (via <cite>fit_encoder()</cite> or <cite>scale=</cite>) for large or high-resolution inputs.</p></li>
</ul>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id19">
<span class="sig-name descname"><span class="pre">Example</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">with</span> <span class="pre">scaling</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id19" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">----------------------</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">s2</span> <span class="pre">=</span> <span class="pre">SinusoidalEncoder(128,</span> <span class="pre">10000)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">a2</span> <span class="pre">=</span> <span class="pre">s2.encode([0,</span> <span class="pre">1,</span> <span class="pre">1,</span> <span class="pre">2,</span> <span class="pre">2,</span> <span class="pre">3,</span> <span class="pre">4,</span> <span class="pre">5,</span> <span class="pre">6,</span> <span class="pre">6,</span> <span class="pre">7,</span> <span class="pre">7,</span> <span class="pre">7,</span> <span class="pre">8,</span> <span class="pre">8,</span> <span class="pre">8,</span> <span class="pre">8,</span> <span class="pre">16,</span> <span class="pre">99,</span> <span class="pre">130],</span> <span class="pre">scale=100)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">s2.decode(a2)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">Output:</span></span></dt>
<dd><dl class="simple">
<dt>[0.0,</dt><dd><p>1.0000000127956146,
1.0000000127956146,
1.999999922004248,
1.999999922004248,
2.999999960015871,
3.9999999840341935,
5.000000076901334,
5.999999935247362,
5.999999935247362,
6.999999909109446,
6.999999909109446,
6.999999909109446,
8.000000084954152,
8.000000084954152,
8.000000084954152,
8.000000084954152,
16.000000431016925,
99.00000148970207,
129.99999823222058]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">Advanced</span> <span class="pre">Example</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">-----------------</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">import</span> <span class="pre">numpy</span> <span class="pre">as</span> <span class="pre">np</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">import</span> <span class="pre">matplotlib.pyplot</span> <span class="pre">as</span> <span class="pre">plt</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">#</span> <span class="pre">1.</span> <span class="pre">Construct</span> <span class="pre">a</span> <span class="pre">test</span> <span class="pre">input</span> <span class="pre">signal</span> <span class="pre">with</span> <span class="pre">smooth</span> <span class="pre">and</span> <span class="pre">jump</span> <span class="pre">segments</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">x_smooth</span> <span class="pre">=</span> <span class="pre">np.linspace(0,</span> <span class="pre">20,</span> <span class="pre">100)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">x_jumps</span> <span class="pre">=</span> <span class="pre">np.array([25,</span> <span class="pre">25,</span> <span class="pre">26,</span> <span class="pre">27,</span> <span class="pre">100,</span> <span class="pre">101,</span> <span class="pre">130])</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">x</span> <span class="pre">=</span> <span class="pre">np.concatenate([x_smooth,</span> <span class="pre">x_jumps])</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">#</span> <span class="pre">2.</span> <span class="pre">Initialize</span> <span class="pre">encoder</span> <span class="pre">with</span> <span class="pre">high</span> <span class="pre">d_model</span> <span class="pre">and</span> <span class="pre">N</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">s</span> <span class="pre">=</span> <span class="pre">SinusoidalEncoder(d_model=128,</span> <span class="pre">N=10000)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">#</span> <span class="pre">3.</span> <span class="pre">Fit</span> <span class="pre">auto-scaling</span> <span class="pre">to</span> <span class="pre">compress</span> <span class="pre">input</span> <span class="pre">into</span> <span class="pre">sinusoidal-friendly</span> <span class="pre">space</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">s.fit_encoder(x,</span> <span class="pre">target_range=10)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">#</span> <span class="pre">4.</span> <span class="pre">Encode</span> <span class="pre">and</span> <span class="pre">decode</span> <span class="pre">using</span> <span class="pre">all</span> <span class="pre">robust</span> <span class="pre">methods</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">a</span> <span class="pre">=</span> <span class="pre">s.encode(x)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">decoded_lsq,</span> <span class="pre">err_lsq</span> <span class="pre">=</span> <span class="pre">s.decode(a,</span> <span class="pre">method='least_squares',</span> <span class="pre">return_error=True)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">decoded_svd,</span> <span class="pre">err_svd</span> <span class="pre">=</span> <span class="pre">s.decode(a,</span> <span class="pre">method='svd',</span> <span class="pre">return_error=True)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">#</span> <span class="pre">5.</span> <span class="pre">Compare</span> <span class="pre">errors</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">true</span> <span class="pre">=</span> <span class="pre">x</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">lsq_error</span> <span class="pre">=</span> <span class="pre">np.abs(decoded_lsq</span> <span class="pre">-</span> <span class="pre">true)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">svd_error</span> <span class="pre">=</span> <span class="pre">np.abs(decoded_svd</span> <span class="pre">-</span> <span class="pre">true)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">#</span> <span class="pre">6.</span> <span class="pre">Plot</span> <span class="pre">results</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">fig,</span> <span class="pre">axs</span> <span class="pre">=</span> <span class="pre">plt.subplots(2,</span> <span class="pre">2,</span> <span class="pre">figsize=(12,</span> <span class="pre">8))</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">axs[0,</span> <span class="pre">0].plot(true,</span> <span class="pre">label=&quot;Original&quot;)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">axs[0,</span> <span class="pre">0].plot(decoded_lsq,</span> <span class="pre">'--',</span> <span class="pre">label=&quot;Decoded</span> <span class="pre">(LSQ)&quot;)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">axs[0,</span> <span class="pre">0].plot(decoded_svd,</span> <span class="pre">':',</span> <span class="pre">label=&quot;Decoded</span> <span class="pre">(SVD)&quot;)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">axs[0,</span> <span class="pre">0].set_title(&quot;Decoded</span> <span class="pre">vs</span> <span class="pre">Original&quot;)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">axs[0,</span> <span class="pre">0].legend()</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">axs[0,</span> <span class="pre">1].plot(lsq_error,</span> <span class="pre">label=&quot;Abs</span> <span class="pre">Error</span> <span class="pre">(LSQ)&quot;)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">axs[0,</span> <span class="pre">1].plot(svd_error,</span> <span class="pre">label=&quot;Abs</span> <span class="pre">Error</span> <span class="pre">(SVD)&quot;)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">axs[0,</span> <span class="pre">1].set_yscale('log')</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">axs[0,</span> <span class="pre">1].set_title(&quot;Absolute</span> <span class="pre">Decoding</span> <span class="pre">Error</span> <span class="pre">(log</span> <span class="pre">scale)&quot;)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">axs[0,</span> <span class="pre">1].legend()</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">axs[1,</span> <span class="pre">0].plot(err_lsq,</span> <span class="pre">label=&quot;Residual</span> <span class="pre">Norm</span> <span class="pre">(LSQ)&quot;)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">axs[1,</span> <span class="pre">0].plot(err_svd,</span> <span class="pre">label=&quot;Residual</span> <span class="pre">Norm</span> <span class="pre">(SVD)&quot;)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">axs[1,</span> <span class="pre">0].set_yscale('log')</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">axs[1,</span> <span class="pre">0].set_title(&quot;Reconstruction</span> <span class="pre">Residuals&quot;)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">axs[1,</span> <span class="pre">0].legend()</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">axs[1,</span> <span class="pre">1].hist(lsq_error,</span> <span class="pre">bins=50,</span> <span class="pre">alpha=0.7,</span> <span class="pre">label=&quot;LSQ&quot;)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">axs[1,</span> <span class="pre">1].hist(svd_error,</span> <span class="pre">bins=50,</span> <span class="pre">alpha=0.5,</span> <span class="pre">label=&quot;SVD&quot;)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">axs[1,</span> <span class="pre">1].set_title(&quot;Histogram</span> <span class="pre">of</span> <span class="pre">Absolute</span> <span class="pre">Errors&quot;)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">axs[1,</span> <span class="pre">1].legend()</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">plt.suptitle(&quot;🌀</span> <span class="pre">SinusoidalEncoder:</span> <span class="pre">Accuracy</span> <span class="pre">Evaluation&quot;,</span> <span class="pre">fontsize=14)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">plt.tight_layout()</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">&gt;&gt;&gt;</span> <span class="pre">plt.show()</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.SinusoidalEncoder.References">
<span class="sig-name descname"><span class="pre">References</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">for</span> <span class="pre">the</span> <span class="pre">encoding</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.SinusoidalEncoder.References" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">------------------------------</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">\*</span> <span class="pre">Vaswani</span> <span class="pre">et</span> <span class="pre">al.</span> <span class="pre">(2017),</span> <span class="pre">&quot;Attention</span> <span class="pre">is</span> <span class="pre">All</span> <span class="pre">You</span> <span class="pre">Need&quot;</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">\*</span> <span class="pre">https://en.wikipedia.org/wiki/Transformer_(deep_learning_architecture)</span></span></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id20">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">angle_difference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">emb</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id20" title="Link to this definition"></a></dt>
<dd><p>Compute angular differences (∆θ) between consecutive embeddings.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>emb</strong> (<em>np.ndarray</em>) – Encoded array of shape (n, 2d)</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Array of shape (n-1, d) of angular differences in radians</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id21">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">complex_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">emb1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">emb2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'L2'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id21" title="Link to this definition"></a></dt>
<dd><p>Compute distance between two encoded arrays using complex projection.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>emb1</strong> (<em>np.ndarray</em>) – Encoded arrays of shape (…, 2d) to compare.</p></li>
<li><p><strong>emb2</strong> (<em>np.ndarray</em>) – Encoded arrays of shape (…, 2d) to compare.</p></li>
<li><p><strong>norm</strong> (<em>str</em>) – ‘L2’ for Euclidean norm, ‘cos’ for cosine angle distance.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Distance values per sample (1D array)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id22">
<span class="sig-name descname"><span class="pre">decode</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">embedding</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'least_squares'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_error</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id22" title="Link to this definition"></a></dt>
<dd><p>Decode sinusoidal embeddings back to original values using selected method.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>embedding</strong> (<em>np.ndarray</em>) – Encoded sinusoidal array of shape (n, d_model)</p></li>
<li><p><strong>method</strong> (<em>str</em>) – Decoding strategy: ‘least_squares’ (default), ‘optimize’, or ‘naive’</p></li>
<li><p><strong>return_error</strong> (<em>bool</em>) – If True, returns (decoded_values, residual_error) as a tuple</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>decoded</strong> (<em>list or array</em>) – Reconstructed input values</p></li>
<li><p><strong>residual</strong> (<em>np.ndarray, optional</em>) – Residuals of decoding (per sample), returned only if return_error=True</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id23">
<span class="sig-name descname"><span class="pre">encode</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id23" title="Link to this definition"></a></dt>
<dd><p>Encode input values into sinusoidal embeddings.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>values</strong> (<em>array-like</em>) – Values to encode.</p></li>
<li><p><strong>scale</strong> (<em>float</em><em> or </em><em>None</em>) – Rescaling factor. If None, auto-scaling is applied if enabled.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Embedded values of shape (n, d_model)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id24">
<span class="sig-name descname"><span class="pre">fit_encoder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10.0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id24" title="Link to this definition"></a></dt>
<dd><p>Fit a scaling factor to normalize values into a target sinusoidal-safe range.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>values</strong> (<em>array-like</em>) – Original values to encode (will determine scale).</p></li>
<li><p><strong>target_range</strong> (<em>float</em>) – Maximum scaled range to span (e.g. [0, 10]).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Recommended scale factor stored internally.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id25">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">group_centroid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">emb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_std</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id25" title="Link to this definition"></a></dt>
<dd><p>Compute the centroid (average embedding) of each group in complex sinusoidal space.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>emb</strong> (<em>np.ndarray</em>) – Encoded array of shape (n, 2d), with sin and cos interleaved.</p></li>
<li><p><strong>labels</strong> (<em>list</em><em> or </em><em>np.ndarray</em><em>, </em><em>optional</em>) – Group labels (n,). If None, the entire set is treated as one group.</p></li>
<li><p><strong>return_std</strong> (<em>bool</em>) – Whether to also return the standard deviation per group.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary mapping each group label to its centroid (2d real array).
If return_std=True, also includes key ‘&lt;label&gt;_std’ with standard deviation.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id26">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">pairwise_similarity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">emb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metric</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cosine'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id26" title="Link to this definition"></a></dt>
<dd><p>Compute a pairwise similarity (or distance) matrix in sinusoidal embedding space.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>emb</strong> (<em>np.ndarray</em>) – Encoded array of shape (n, 2d).</p></li>
<li><p><strong>metric</strong> (<em>str</em>) – Distance metric: ‘cosine’ for 1 - cosine similarity, ‘L2’ for Euclidean norm.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Pairwise similarity matrix of shape (n, n)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id27">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">phase_alignment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">emb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id27" title="Link to this definition"></a></dt>
<dd><p>Align embedding <cite>emb</cite> to reference <cite>ref</cite> using complex phase.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>emb</strong> (<em>np.ndarray</em>) – Encoded array to align (n, 2d)</p></li>
<li><p><strong>ref</strong> (<em>np.ndarray</em>) – Reference encoded array (n, 2d)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Aligned encoding of <cite>emb</cite>, same shape as input.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id28">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">phase_unwrap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">emb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id28" title="Link to this definition"></a></dt>
<dd><p>Perform Fourier-like phase unwrapping on sinusoidal embedding.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>emb</strong> (<em>np.ndarray</em>) – Encoded array of shape (n, 2d).</p></li>
<li><p><strong>normalize</strong> (<em>bool</em>) – Whether to scale unwrapped phases to [0, 1].</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Phase unwrapped matrix of shape (n, d), optionally normalized.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id29">
<span class="sig-name descname"><span class="pre">set_decode_tolerance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.001</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id29" title="Link to this definition"></a></dt>
<dd><p>Set maximum acceptable residual error for decoding.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>tol</strong> (<em>float</em>) – Residual threshold above which a warning is triggered.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id30">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sindecode_dna_grouped</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">grouped</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_code</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">d_part</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">32</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">N</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id30" title="Link to this definition"></a></dt>
<dd><p>Decode sinusoidal embeddings grouped by letter into symbolic segments.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>grouped</strong> (<em>dict</em>) – Dictionary of letter: embeddings</p></li>
<li><p><strong>reference_code</strong> (<em>dict</em>) – Must include ‘dx’ (sampling resolution).</p></li>
<li><p><strong>d_part</strong> (<em>int</em>) – Number of dimensions per part (start, width, height)</p></li>
<li><p><strong>N</strong> (<em>int</em>) – Frequency base</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary with keys: letters, widths, heights, xloc, iloc, dx</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id31">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sinencode_dna_grouped</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">code</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">d_part</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">32</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">N</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id31" title="Link to this definition"></a></dt>
<dd><p>Encode symbolic code segments grouped by letter into sinusoidal embeddings.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>code</strong> (<em>dict</em>) – <dl class="simple">
<dt>Must contain:</dt><dd><ul>
<li><p>’letters’: str</p></li>
<li><p>’xloc’: list of (start, end) tuples</p></li>
<li><p>’widths’: list of float</p></li>
<li><p>’heights’: list of float</p></li>
</ul>
</dd>
</dl>
</p></li>
<li><p><strong>d_part</strong> (<em>int</em>) – Number of dimensions per field (position, width, height)</p></li>
<li><p><strong>N</strong> (<em>int</em>) – Sinusoidal frequency base</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary of embeddings by letter: {letter: np.ndarray(n, 3*d_part)}</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id32">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">to_complex</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">emb</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id32" title="Link to this definition"></a></dt>
<dd><p>Convert sinusoidal embedding into a complex array using Euler’s identity.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>emb</strong> (<em>np.ndarray</em>) – Array of shape (…, 2d) where sin/cos pairs are stored.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Complex array of shape (…, d) with values exp(i * theta)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="id33">
<span class="sig-name descname"><span class="pre">verify_roundtrip</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'least_squares'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'auto'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_details</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#id33" title="Link to this definition"></a></dt>
<dd><p>Perform an encode → decode → compare roundtrip and report accuracy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>values</strong> (<em>array-like</em>) – Original values to test.</p></li>
<li><p><strong>method</strong> (<em>str</em>) – Decoding method: ‘least_squares’, ‘svd’, ‘optimize’, or ‘naive’.</p></li>
<li><p><strong>scale</strong> (<em>float</em><em> or </em><em>'auto'</em><em> or </em><em>None</em>) – Scaling strategy: ‘auto’ uses fit_encoder(), float uses fixed scaling, None disables scaling.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em>) – If True, prints accuracy report.</p></li>
<li><p><strong>return_details</strong> (<em>bool</em>) – If True, also returns the encoded array, decoded values, and residuals.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>success</strong> (<em>bool</em>) – Whether all values were accurately recovered within residual tolerance.</p></li>
<li><p><strong>details</strong> (<em>tuple, optional</em>) – Tuple (encoded, decoded, residuals) if return_details=True</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sig2dna_core.signomics.generator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sig2dna_core.signomics.</span></span><span class="sig-name descname"><span class="pre">generator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kind</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gauss'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.generator" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sig2dna_core.signomics.import_local_module">
<span class="sig-prename descclassname"><span class="pre">sig2dna_core.signomics.</span></span><span class="sig-name descname"><span class="pre">import_local_module</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">relative_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.import_local_module" title="Link to this definition"></a></dt>
<dd><p>Import a module by name from a file path relative to the calling module.</p>
<section id="usage-here">
<h2>Usage here:<a class="headerlink" href="#usage-here" title="Link to this heading"></a></h2>
<p>import_local_module(“figprint”, “figprint.py”) replaces import figprint (zero-installation)</p>
</section>
<section id="parameters">
<h2>Parameters:<a class="headerlink" href="#parameters" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>name<span class="classifier">str</span></dt><dd><p>Name to assign to the module (used internally).</p>
</dd>
<dt>relative_path<span class="classifier">str</span></dt><dd><p>Path relative to the calling module’s location.</p>
</dd>
</dl>
</section>
<section id="returns">
<h2>Returns:<a class="headerlink" href="#returns" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>module</dt><dd><p>Imported module object.</p>
</dd>
</dl>
</section>
<section id="id34">
<h2>Example:<a class="headerlink" href="#id34" title="Link to this heading"></a></h2>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">figprint</span> <span class="o">=</span> <span class="n">import_local_module</span><span class="p">(</span><span class="s2">&quot;figprint&quot;</span><span class="p">,</span> <span class="s2">&quot;figprint.py&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">figprint</span><span class="o">.</span><span class="n">print_pdf</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>
</pre></div>
</div>
</section>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sig2dna_core.signomics.peaks">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sig2dna_core.signomics.</span></span><span class="sig-name descname"><span class="pre">peaks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.peaks" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A class for managing a collection of peak definitions used in synthetic signal generation.</p>
<p>Each peak is represented as a dictionary with the following fields:
- ‘name’ (str): unique identifier (autogenerated if not provided)
- ‘x’ (float): center position (e.g., time, wavenumber, index)
- ‘w’ (float): width (related to FWHM)
- ‘h’ (float): peak height
- ‘type’ (str): generator type (e.g., ‘gauss’, ‘lorentz’, ‘triangle’)</p>
<p>Supports:
- Flexible addition and broadcasting of peak parameters
- Named or indexed access to individual or multiple peaks
- Overloaded operators for peak translation and scaling
- Utility methods: update, sort, rename, remove_duplicates, copy
- Conversion to signal object via <cite>.to_signal()</cite>
- Informative __str__ and __repr__ output</p>
<p>This class is used to build reproducible and structured test cases for symbolic encoding (e.g., sig2dna).</p>
<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.peaks.add">
<span class="sig-name descname"><span class="pre">add</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">w</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">h</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gauss'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.peaks.add" title="Link to this definition"></a></dt>
<dd><p>Add one or multiple peaks to the collection.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<em>float</em><em> or </em><em>array-like</em>) – Center positions of the peaks.</p></li>
<li><p><strong>w</strong> (<em>float</em><em> or </em><em>array-like</em>) – Width(s) of the peaks (broadcastable).</p></li>
<li><p><strong>h</strong> (<em>float</em><em> or </em><em>array-like</em>) – Height(s) of the peaks (broadcastable).</p></li>
<li><p><strong>name</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em><em> or </em><em>None</em>) – Peak name(s); auto-generated if None or duplicate.</p></li>
<li><p><strong>type</strong> (<em>str</em>) – Generator type, e.g., ‘gauss’, ‘lorentz’, etc.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.peaks.as_dict">
<span class="sig-name descname"><span class="pre">as_dict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.peaks.as_dict" title="Link to this definition"></a></dt>
<dd><p>Return the list of peaks as dict</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.peaks.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.peaks.copy" title="Link to this definition"></a></dt>
<dd><p>Return a deep-copy of the peaks</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.peaks.names">
<span class="sig-name descname"><span class="pre">names</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.peaks.names" title="Link to this definition"></a></dt>
<dd><p>Return the list of names</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.peaks.remove_duplicates">
<span class="sig-name descname"><span class="pre">remove_duplicates</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.peaks.remove_duplicates" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.peaks.rename">
<span class="sig-name descname"><span class="pre">rename</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'P'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.peaks.rename" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.peaks.sort">
<span class="sig-name descname"><span class="pre">sort</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'asc'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.peaks.sort" title="Link to this definition"></a></dt>
<dd><p>Sort peaks in-place based on their center positions (x values).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>order</strong> (<em>str</em>) – <dl class="simple">
<dt>Sorting direction. Use:</dt><dd><ul class="simple">
<li><p>”asc” for ascending (default)</p></li>
<li><p>”desc” for descending</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.peaks.to_signal">
<span class="sig-name descname"><span class="pre">to_signal</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">generator_map</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x0</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.peaks.to_signal" title="Link to this definition"></a></dt>
<dd><p>Generate a signal from a peaks object. Optionally restrict to a subset.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.peaks.update">
<span class="sig-name descname"><span class="pre">update</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.peaks.update" title="Link to this definition"></a></dt>
<dd><p>Update or insert peaks from a list of dictionaries.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>data</strong> (<em>list</em><em> of </em><em>dict</em>) – Each dict must include at least ‘x’, ‘w’, ‘h’.
If ‘name’ matches an existing peak, it will be updated.
If ‘name’ is new or missing, the peak is appended.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sig2dna_core.signomics.</span></span><span class="sig-name descname"><span class="pre">signal</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'signal'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'generic'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'index'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_unit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'-'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'intensity'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_unit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'a.u.'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'array'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">user</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">date</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cwd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linestyle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'-'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">message</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fullhistory</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.signal" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>signal: A self-documented 1D analytical signal container for reproducible scientific workflows.</p>
<p>This class is designed for lab-grade signal processing and traceable data storage.
It represents a discrete 1D signal (e.g., chromatogram, spectrum, transient) with full metadata,
support for symbolic transformation, numerical operations, plotting, and structured saving/loading.</p>
<p>Key features include:
- Portable metadata (user, time, host, cwd, version)
- Domain-aware plots and operations
- Reproducible signal serialization in JSON or compressed format
- Full traceability of all transformation events
- Optional recursive backup of prior states</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal.x">
<span class="sig-name descname"><span class="pre">x</span></span><a class="headerlink" href="#sig2dna_core.signomics.signal.x" title="Link to this definition"></a></dt>
<dd><p>Sampling domain (e.g., time, wavelength, chemical shift).</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal.y">
<span class="sig-name descname"><span class="pre">y</span></span><a class="headerlink" href="#sig2dna_core.signomics.signal.y" title="Link to this definition"></a></dt>
<dd><p>Signal values aligned with x.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal.name">
<span class="sig-name descname"><span class="pre">name</span></span><a class="headerlink" href="#sig2dna_core.signomics.signal.name" title="Link to this definition"></a></dt>
<dd><p>Label for plots and file storage (used as default filename).</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal.type">
<span class="sig-name descname"><span class="pre">type</span></span><a class="headerlink" href="#sig2dna_core.signomics.signal.type" title="Link to this definition"></a></dt>
<dd><p>Optional tag (e.g., ‘GC-MS’, ‘FTIR’, ‘NMR’, ‘synthetic’).</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal.x_label">
<span class="sig-name descname"><span class="pre">x_label</span></span><a class="headerlink" href="#sig2dna_core.signomics.signal.x_label" title="Link to this definition"></a></dt>
<dd><p>Label for the x-axis (e.g., ‘wavenumber’).</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal.x_unit">
<span class="sig-name descname"><span class="pre">x_unit</span></span><a class="headerlink" href="#sig2dna_core.signomics.signal.x_unit" title="Link to this definition"></a></dt>
<dd><p>Unit of the x-axis (e.g., ‘cm⁻¹’).</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal.source">
<span class="sig-name descname"><span class="pre">source</span></span><a class="headerlink" href="#sig2dna_core.signomics.signal.source" title="Link to this definition"></a></dt>
<dd><p>Origin label (‘array’, ‘peaks’, ‘noise’, ‘imported’…).</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal.metadata">
<span class="sig-name descname"><span class="pre">metadata</span></span><a class="headerlink" href="#sig2dna_core.signomics.signal.metadata" title="Link to this definition"></a></dt>
<dd><p>Includes user, date, host, cwd, version — filled automatically unless overridden.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal.color">
<span class="sig-name descname"><span class="pre">color</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">str</span> <span class="pre">or</span> <span class="pre">[rgb])</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linestyle</span> <span class="pre">(str)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linewidth</span> <span class="pre">(str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.signal.color" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal._previous">
<span class="sig-name descname"><span class="pre">_previous</span></span><a class="headerlink" href="#sig2dna_core.signomics.signal._previous" title="Link to this definition"></a></dt>
<dd><p>deep-copy of current object</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sig2dna_core.signomics.signal" title="sig2dna_core.signomics.signal">signal</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal._history">
<span class="sig-name descname"><span class="pre">_history</span></span><a class="headerlink" href="#sig2dna_core.signomics.signal._history" title="Link to this definition"></a></dt>
<dd><p>“<a class="reference external" href="mailto:user&#37;&#52;&#48;host">user<span>&#64;</span>host</a>:timestamp | uidkey” :{“action”:str, “details”: str}</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">Key</span> <span class="pre">Methods</span></span></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">-----------</span></span></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">-</span> <span class="pre">from_peaks(...)</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Construct signal from a <cite>peaks</cite> object</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">-</span> <span class="pre">add_noise(...)</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Return noisy variant (Poisson, Gaussian, ramp or constant bias)</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">-</span> <span class="pre">align_with(...)</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Align this signal with another (same x domain)</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">-</span> <span class="pre">copy()</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Deep copy</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">-</span> <span class="pre">save(...)</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Save as JSON or .gz (optional CSV export)</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">-</span> <span class="pre">load(...)</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Load from saved file</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">-</span> <span class="pre">plot(...)</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Plot the signal with axis labels</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">-</span> <span class="pre">backup(...)</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Backup current signal (deep-copy stored in _previous)</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">-</span> <span class="pre">restore(...)</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Restore the previous state of the signal</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">-</span> <span class="pre">apply_poisson_baseline_filter(...)</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Apply a Poisson-based filter</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">-</span> <span class="pre">enable_fullhistory</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>enable full history</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">-</span> <span class="pre">disable_fullhistory</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>disable full history</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">-</span> <span class="pre">_toDNA(signal)</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sig2dna_core.signomics.DNAsignal" title="sig2dna_core.signomics.DNAsignal">DNAsignal</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">Overloaded</span> <span class="pre">Operators</span></span></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">--------------------</span></span></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">-</span> <span class="pre">+,</span> <span class="pre">-,</span> <span class="pre">\*,</span> <span class="pre">/</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Operates on signals or scalars, aligns if needed</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">-</span> <span class="pre">+=,</span> <span class="pre">-=,</span> <span class="pre">\*=,</span> <span class="pre">/=</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>In-place functional versions (returns new signal)</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">Low-level</span> <span class="pre">Methods</span></span></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">-----------------</span></span></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">-</span> <span class="pre">_current_stamp()</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>stamp for events (static method)</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">-</span> <span class="pre">_copystatic()</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>deep-copy of signal only (use copy for a full copy) (static method)</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">-</span> <span class="pre">_events()</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>register a processing step</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">-</span> <span class="pre">_to_serializable</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Convert the signal into a dictionary suitable for JSON export</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">-</span> <span class="pre">_from_serizalizable</span></span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>convert a dict (e.g., from JSON import) to signal</p>
</dd>
</dl>
</dd></dl>

<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">signal</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;sample&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;FTIR&quot;</span><span class="p">,</span> <span class="n">x_label</span><span class="o">=</span><span class="s2">&quot;wavenumber&quot;</span><span class="p">,</span> <span class="n">x_unit</span><span class="o">=</span><span class="s2">&quot;cm⁻¹&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">.</span><span class="n">add_noise</span><span class="p">(</span><span class="s2">&quot;gaussian&quot;</span><span class="p">,</span> <span class="mf">0.05</span><span class="p">)</span><span class="o">.</span><span class="n">plot</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">.</span><span class="n">save</span><span class="p">()</span>  <span class="c1"># saves to ./sample.json.gz</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s2</span> <span class="o">=</span> <span class="n">signal</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;sample.json.gz&quot;</span><span class="p">)</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal.add_noise">
<span class="sig-name descname"><span class="pre">add_noise</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kind</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gaussian'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bias</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.signal.add_noise" title="Link to this definition"></a></dt>
<dd><p>Return a new signal with noise and/or bias added.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal.align_with">
<span class="sig-name descname"><span class="pre">align_with</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'union'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.signal.align_with" title="Link to this definition"></a></dt>
<dd><p>Align two signals to a common x grid with interpolation and padding.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="#sig2dna_core.signomics.signal" title="sig2dna_core.signomics.signal"><em>signal</em></a>) – the other signal to align with</p></li>
<li><p><strong>mode</strong> (<em>str</em>) – ‘union’ (default) or ‘intersection’</p></li>
<li><p><strong>n</strong> (<em>int</em>) – number of points for the new grid</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(self_interp, other_interp) as new signal instances</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal.apply_poisson_baseline_filter">
<span class="sig-name descname"><span class="pre">apply_poisson_baseline_filter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">window_ratio</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.02</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gain</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proba</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.9</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.signal.apply_poisson_baseline_filter" title="Link to this definition"></a></dt>
<dd><blockquote>
<div><p>Apply a baseline filter assuming Poisson-dominated statistics with adjustable gain
and a rejection threshold based on the Bienaymé-Tchebychev inequality.</p>
<p>The signal is filtered by removing values likely caused by statistical noise
(false peaks) using a per-point threshold defined from local statistics:</p>
<ul class="simple">
<li><p>Local mean: $$ mu_t =</p></li>
</ul>
</div></blockquote>
<dl class="simple">
<dt>rac{1}{w} sum_{i in W(t)} y_i $$</dt><dd><ul class="simple">
<li><p>Local std dev: $$ sigma_t = sqrt{mu_t cdot        ext{gain}} $$</p></li>
<li><p>Coefficient of variation: $$  ext{cv}_t =</p></li>
</ul>
</dd>
<dt>rac{sigma_t}{mu_t} $$</dt><dd><ul class="simple">
<li><p>Estimated local intensity (lambda): $$ lambda_t =</p></li>
</ul>
</dd>
<dt>rac{1}{   ext{cv}_t^2} $$</dt><dd><ul class="simple">
<li><p>Bienaymé-Tchebychev threshold: $$     ext{threshold}_t =</p></li>
</ul>
</dd>
</dl>
<p>rac{1}{sqrt{1 - p}} cdot sqrt{10 lambda_t cdot Delta t} $$</p>
<blockquote>
<div><dl class="simple">
<dt>window_ratio<span class="classifier">float, default=0.02</span></dt><dd><p>Ratio of signal length used as window size (must yield odd integer ≥ 11).</p>
</dd>
<dt>gain<span class="classifier">float, default=1.0</span></dt><dd><p>Linear amplification factor applied to simulate signal counts.</p>
</dd>
<dt>proba<span class="classifier">float, default=0.9</span></dt><dd><p>Minimum probability to consider a signal point significant. Must be in (0, 1).</p>
</dd>
</dl>
<dl class="simple">
<dt>signal</dt><dd><p>The current signal instance (self), with updated <cite>y</cite>.</p>
</dd>
</dl>
<dl class="simple">
<dt>ValueError</dt><dd><p>If the window size is too small for reliable statistics.</p>
</dd>
</dl>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal.backup">
<span class="sig-name descname"><span class="pre">backup</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fullhistory</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">message</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.signal.backup" title="Link to this definition"></a></dt>
<dd><p>Backup current state in _previous</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.signal.copy" title="Link to this definition"></a></dt>
<dd><p>Deep copy of the signal, excluding full history control flag</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal.disable_fullhistory">
<span class="sig-name descname"><span class="pre">disable_fullhistory</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.signal.disable_fullhistory" title="Link to this definition"></a></dt>
<dd><p>Disable full history tracking</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal.enable_fullhistory">
<span class="sig-name descname"><span class="pre">enable_fullhistory</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.signal.enable_fullhistory" title="Link to this definition"></a></dt>
<dd><p>Enable full history tracking</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal.from_peaks">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_peaks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peaks_obj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">generator_map</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'from_peaks'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x0</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.signal.from_peaks" title="Link to this definition"></a></dt>
<dd><p>Generate a signal from a set of peaks.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>peaks_obj</strong> (<a class="reference internal" href="#sig2dna_core.signomics.peaks" title="sig2dna_core.signomics.peaks"><em>peaks</em></a>) – A list-like object containing peak definitions.</p></li>
<li><p><strong>x</strong> (<em>array-like</em><em>, </em><em>float</em><em>, or </em><em>None</em>) – If None: compute x domain from peaks.
If scalar: interpreted as xmax; linspace from x0 to xmax.
If array: use as x directly.</p></li>
<li><p><strong>generator_map</strong> (<em>dict</em><em> or </em><em>None</em>) – Optional map of peak type → generator instance (default is Gaussian).</p></li>
<li><p><strong>name</strong> (<em>str</em>) – Name of the signal instance.</p></li>
<li><p><strong>x0</strong> (<em>float</em><em> or </em><em>None</em>) – Left bound of the domain (used only if x is None or scalar).
If None: inferred from peaks.</p></li>
<li><p><strong>n</strong> (<em>int</em>) – Number of points in the generated x array.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A new signal instance generated from the peaks.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sig2dna_core.signomics.signal" title="sig2dna_core.signomics.signal">signal</a></p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>p = peaks()
p.add(x=[400, 800, 1600], w=30, h=[1.0, 0.6, 0.9], type=”gauss”)
s = signal.from_peaks(p, x0=300, n=2048)
s.plot()</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal.load">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filepath</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.signal.load" title="Link to this definition"></a></dt>
<dd><p>Load a signal from a JSON or gzipped JSON file, including recursive _previous.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>filepath</strong> (<em>str</em><em> or </em><em>Path</em>) – Path to the JSON or .gz file</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A fully reconstructed signal object</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sig2dna_core.signomics.signal" title="sig2dna_core.signomics.signal">signal</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal.n">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">n</span></span><a class="headerlink" href="#sig2dna_core.signomics.signal.n" title="Link to this definition"></a></dt>
<dd><p>Return the length of the signal and None if it is None</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal.plot">
<span class="sig-name descname"><span class="pre">plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linestyle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fontsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">12</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">newfig</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.signal.plot" title="Link to this definition"></a></dt>
<dd><p>Plot the signal using matplotlib, applying either internal style settings
or overrides provided at call time.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em><em>, </em><em>optional</em>) – Axis to plot on. If None, uses current axis or new figure if newfig=True.</p></li>
<li><p><strong>label</strong> (<em>str</em><em>, </em><em>optional</em>) – Legend label. Defaults to self.name.</p></li>
<li><p><strong>color</strong> (<em>str</em><em> or </em><em>None</em>) – Line color. If None, uses default matplotlib cycling.</p></li>
<li><p><strong>linestyle</strong> (<em>str</em><em> or </em><em>None</em>) – Line style (e.g., ‘-’, ‘–‘). If None, uses self.linestyle.</p></li>
<li><p><strong>linewidth</strong> (<em>float</em><em> or </em><em>None</em>) – Line width. If None, uses self.linewidth.</p></li>
<li><p><strong>fontsize</strong> (<em>int</em><em> or </em><em>str</em>) – Font size for axis labels and legend. Can use values like ‘small’, ‘large’.</p></li>
<li><p><strong>newfig</strong> (<em>bool</em>) – If True, creates a new figure before plotting.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>matplotlib.figure.Figure</em></p></li>
<li><p><em>matplotlib.axes.Axes</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal.restore">
<span class="sig-name descname"><span class="pre">restore</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.signal.restore" title="Link to this definition"></a></dt>
<dd><p>Restore the previous signal version if available</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal.sample">
<span class="sig-name descname"><span class="pre">sample</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x_new</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.signal.sample" title="Link to this definition"></a></dt>
<dd><p>Interpolate values from x</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal.save">
<span class="sig-name descname"><span class="pre">save</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filepath</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zip</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">export_csv</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.signal.save" title="Link to this definition"></a></dt>
<dd><p>Save signal to JSON (optionally compressed) and optionally CSV.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filepath</strong> (<em>str</em><em> or </em><em>Path</em><em> or </em><em>None</em>) – If None, builds path from metadata[‘cwd’] and self.name + ‘.json[.gz]’.
If a directory, appends name + ‘.json[.gz]’.
If a file, uses as is.</p></li>
<li><p><strong>zip</strong> (<em>bool</em>) – Whether to compress the JSON file using gzip. Default: True.</p></li>
<li><p><strong>export_csv</strong> (<em>bool</em>) – If True, also save a .csv file (x,y) alongside the JSON.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal_collection">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sig2dna_core.signomics.</span></span><span class="sig-name descname"><span class="pre">signal_collection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">signals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1024</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'union'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.signal_collection" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></p>
<p>A container class for multiple <cite>signal</cite> instances that ensures alignment on a shared x-grid.</p>
<p>The collection is used to manage, compare, combine, or visualize multiple signals (e.g., from
replicates, experiments, synthetic scenarios). Signals are interpolated and padded on insertion
so all have the same shape and domain. Arithmetic, matrix extraction, and overlay plots are supported.</p>
<section id="id35">
<h2>Parameters:<a class="headerlink" href="#id35" title="Link to this heading"></a></h2>
<dl class="simple">
<dt><a href="#id36"><span class="problematic" id="id37">*</span></a>signals<span class="classifier">signal</span></dt><dd><p>One or more signal instances to include (they are copied and aligned).</p>
</dd>
<dt>n<span class="classifier">int</span></dt><dd><p>Number of sampling points in the aligned x grid (default: 1000).</p>
</dd>
<dt>mode<span class="classifier">str</span></dt><dd><p>Alignment mode: ‘union’ or ‘intersection’ of x-ranges.</p>
</dd>
</dl>
</section>
<section id="core-attributes">
<h2>Core Attributes:<a class="headerlink" href="#core-attributes" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>mode<span class="classifier">str</span></dt><dd><p>Alignment strategy used (“union” or “intersection”).</p>
</dd>
<dt>n<span class="classifier">int</span></dt><dd><p>Number of x-points used in alignment (default=1024).</p>
</dd>
</dl>
</section>
<section id="id38">
<h2>Key Methods:<a class="headerlink" href="#id38" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>append(signal)              → add and align a new signal</p></li>
<li><p>to_matrix()                 → convert signals to a 2D array (n_signals x n_points)</p></li>
<li><p>mean(coeffs=None)           → weighted or unweighted mean</p></li>
<li><p>sum(coeffs=None)            → weighted or unweighted sum</p></li>
<li><p>plot(…)                   → overlay signals with optional mean/sum</p></li>
<li><p>copy                        → all signals stored are deep copies</p></li>
<li><p>generate_synthetic          → signal collection composed of random peaks.</p></li>
<li><p>__getitem__(…)            → slice, list, or name-based access to signals</p></li>
<li><p>__repr__ / __str__          → report contents with span and names</p></li>
<li><p>_toDNA(signal_collection)   → list of DNAsignals</p></li>
</ul>
</section>
<section id="access-patterns">
<h2>Access Patterns:<a class="headerlink" href="#access-patterns" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>sc[0:3]           → subcollection by slice</p></li>
<li><p>sc[[0, 2]]        → subcollection by list of indices</p></li>
<li><p>sc[“name”]        → return a copy of signal with that name</p></li>
<li><p>sc[“A”, “B”]      → return a subcollection with those names</p></li>
</ul>
</section>
<section id="examples">
<h2>Examples:<a class="headerlink" href="#examples" title="Link to this heading"></a></h2>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span> <span class="o">=</span> <span class="n">signal_collection</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span> <span class="n">s2</span><span class="p">,</span> <span class="n">s3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">show_mean</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span>         <span class="c1"># sub-collection (copy)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="p">[</span><span class="s2">&quot;peak1&quot;</span><span class="p">]</span>     <span class="c1"># get copy of signal named &#39;peak1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mat</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">to_matrix</span><span class="p">()</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span><span class="o">.</span><span class="n">plot</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">coeffs</span><span class="o">=</span><span class="p">[</span><span class="mf">0.4</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">])</span><span class="o">.</span><span class="n">plot</span><span class="p">()</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal_collection.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">new_signal</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.signal_collection.append" title="Link to this definition"></a></dt>
<dd><p>Append and align the new signal to the existing collection.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal_collection.generate_mixtures">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">generate_mixtures</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_mixtures</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_peaks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">16</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">peaks_per_mixture</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(3,</span> <span class="pre">8)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">amplitude_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0.5,</span> <span class="pre">2)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flatten</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'mean'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_signals</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_peaks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kinds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('gauss',)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0.5,</span> <span class="pre">3)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">height_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(1.0,</span> <span class="pre">5.0)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0,</span> <span class="pre">500)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_points</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1024</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.signal_collection.generate_mixtures" title="Link to this definition"></a></dt>
<dd><p>Generate synthetic mixtures of signals by combining a subset of base peaks.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n_mixtures</strong> (<em>int</em>) – Number of synthetic mixtures to generate.</p></li>
<li><p><strong>max_peaks</strong> (<em>int</em>) – Maximum number of base signals (from which peaks are taken).</p></li>
<li><p><strong>peaks_per_mixture</strong> (<em>tuple</em><em> of </em><em>(</em><em>int</em><em>, </em><em>int</em><em>)</em>) – Range (min, max) for the number of peaks to combine in each mixture.
Cannot exceed <cite>max_peaks</cite>.</p></li>
<li><p><strong>amplitude_range</strong> (<em>tuple</em><em> of </em><em>(</em><em>float</em><em>, </em><em>float</em><em>)</em>) – Random scaling range applied to peak amplitudes in each mixture.</p></li>
<li><p><strong>flatten</strong> (<em>{'sum'</em><em>, </em><em>'mean'}</em><em>, </em><em>default='mean'</em>) – How to combine the signals for each mixture.</p></li>
<li><p><strong>**kwargs</strong> (<em>dict</em>) – All other keyword arguments passed to <cite>generate_synthetic</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>result_collection</strong> (<em>signal_collection</em>) – A collection of synthetic mixed signals.</p></li>
<li><p><strong>all_peaks</strong> (<em>list of dict</em>) – All individual peaks originally generated.</p></li>
<li><p><strong>used_peak_ids</strong> (<em>list of list of str</em>) – For each mixture, the list of peak names used.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>S, pS = signal_collection.generate_mixtures(
…     n_mixtures=30,
…     max_peaks=12,
…     peaks_per_mixture=(4, 8),
…     amplitude_range=(0.2, 1.5),
…     n_signals=12,
…     kinds=(“gauss”,),
…     width_range=(0.5, 3),
…     height_range=(1.0, 5.0),
…     x_range=(0, 500),
…     n_points=2048,
…     normalize=False,
…     seed=123
… )
&gt;&gt;&gt; S.plot()</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal_collection.generate_synthetic">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">generate_synthetic</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_signals</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_peaks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kind_distribution</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'uniform'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kinds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('gauss',</span> <span class="pre">'lorentz',</span> <span class="pre">'triangle')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0,</span> <span class="pre">1000)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_points</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1024</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">avoid_overlap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(20,</span> <span class="pre">60)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">height_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0.5,</span> <span class="pre">1.0)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">noise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bias</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name_prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'synthetic'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.signal_collection.generate_synthetic" title="Link to this definition"></a></dt>
<dd><p>Generate a synthetic signal collection composed of random peaks.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n_signals</strong> (<em>int</em>) – Number of synthetic signals to generate.</p></li>
<li><p><strong>n_peaks</strong> (<em>int</em>) – Number of peaks per signal.</p></li>
<li><p><strong>kind_distribution</strong> (<em>str</em>) – ‘uniform’ → use all peak kinds equally; ‘random’ → random draw from kinds.</p></li>
<li><p><strong>kinds</strong> (<em>tuple</em><em>[</em><em>str</em><em>]</em>) – Generator types to choose from (‘gauss’, ‘lorentz’, ‘triangle’).</p></li>
<li><p><strong>x_range</strong> (<em>tuple</em><em>[</em><em>float</em><em>, </em><em>float</em><em>]</em>) – Start and end of x-domain.</p></li>
<li><p><strong>n_points</strong> (<em>int</em>) – Number of sampling points for each signal (default: 1024).</p></li>
<li><p><strong>avoid_overlap</strong> (<em>bool</em>) – Prevent peaks from overlapping by checking spacing vs. width.</p></li>
<li><p><strong>width_range</strong> (<em>tuple</em><em>[</em><em>float</em><em>, </em><em>float</em><em>]</em>) – Range of widths for the peaks.</p></li>
<li><p><strong>height_range</strong> (<em>tuple</em><em>[</em><em>float</em><em>, </em><em>float</em><em>]</em>) – Range of peak heights.</p></li>
<li><p><strong>normalize</strong> (<em>bool</em>) – Normalize each signal so the highest peak has intensity 1.</p></li>
<li><p><strong>noise</strong> (<em>dict</em><em> or </em><em>None</em>) – Optional noise model, e.g. {“kind”: “gaussian”, “scale”: 0.01}.</p></li>
<li><p><strong>bias</strong> (<em>float</em><em>, </em><em>str</em><em>, </em><em>np.ndarray</em><em>, or </em><a class="reference internal" href="#sig2dna_core.signomics.signal" title="sig2dna_core.signomics.signal"><em>signal</em></a>) – Optional signal bias: can be a constant, ‘ramp’, or signal.</p></li>
<li><p><strong>name_prefix</strong> (<em>str</em>) – Base name for each generated signal.</p></li>
<li><p><strong>seed</strong> (<em>int</em><em> or </em><em>None</em>) – Random seed for reproducibility.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A collection of generated signals.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sig2dna_core.signomics.signal_collection" title="sig2dna_core.signomics.signal_collection">signal_collection</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p># 1. Default random peaks, Gaussian + ramp bias
sc = signal_collection.generate_synthetic(</p>
<blockquote>
<div><p>n_signals=5,
n_peaks=6,
kinds=(“gauss”, “lorentz”, “triangle”),
noise={“kind”: “gaussian”, “scale”: 0.02},
bias=”ramp”,
name_prefix=”test”</p>
</div></blockquote>
<p>)
sc.plot(show_mean=True, fontsize=”large”)</p>
<p># 2. High-res signal, fixed width and height
sc2 = signal_collection.generate_synthetic(</p>
<blockquote>
<div><p>n_signals=3,
n_peaks=8,
kinds=(“gauss”,),
width_range=(30, 30),
height_range=(1.0, 1.0),
x_range=(0, 500),
n_points=2048,
normalize=False,
seed=123</p>
</div></blockquote>
<p>)
sc2.plot(fontsize=14)</p>
<p># 3. Poisson noise, no overlap, save output
sc3 = signal_collection.generate_synthetic(</p>
<blockquote>
<div><p>n_signals=2,
n_peaks=5,
noise={“kind”: “poisson”, “scale”: 2.0},
name_prefix=”poisson_example”</p>
</div></blockquote>
<p>)
for s in sc3:</p>
<blockquote>
<div><p>s.save(export_csv=True)</p>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal_collection.mean">
<span class="sig-name descname"><span class="pre">mean</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">indices_or_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coeffs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.signal_collection.mean" title="Link to this definition"></a></dt>
<dd><p>Mean of selected signals, optionally weighted.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>indices_or_names</strong> (<em>list</em><em>[</em><em>int</em><em> or </em><em>str</em><em>]</em><em>, </em><em>optional</em>) – Signal names or indices to include.</p></li>
<li><p><strong>coeffs</strong> (<em>list</em><em>[</em><em>float</em><em>]</em><em>, </em><em>optional</em>) – Weights for selected signals.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Averaged signal.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sig2dna_core.signomics.signal" title="sig2dna_core.signomics.signal">signal</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal_collection.plot">
<span class="sig-name descname"><span class="pre">plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">indices</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">newfig</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_mean</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_sum</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coeffs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fontsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">12</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colormap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.signal_collection.plot" title="Link to this definition"></a></dt>
<dd><p>Plot selected signals with style attributes and optional overlays.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>indices</strong> (<em>list</em><em>[</em><em>int</em><em>] or </em><em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – Signals to plot by index or name.</p></li>
<li><p><strong>labels</strong> (<em>bool</em>) – Whether to show signal labels.</p></li>
<li><p><strong>title</strong> (<em>str</em>) – Plot title.</p></li>
<li><p><strong>newfig</strong> (<em>bool</em><em> or </em><em>None</em>) – If True, always open a new figure.
If False, use current axes.
If None, open new figure only the first time this collection is plotted.</p></li>
<li><p><strong>ax</strong> (<em>matplotlib axis</em><em>, </em><em>optional</em>) – Axis to draw on.</p></li>
<li><p><strong>show_mean</strong> (<em>bool</em>) – Overlay mean curve.</p></li>
<li><p><strong>show_sum</strong> (<em>bool</em>) – Overlay sum curve.</p></li>
<li><p><strong>coeffs</strong> (<em>list</em><em>[</em><em>float</em><em>]</em><em>, </em><em>optional</em>) – Optional weights for mean/sum.</p></li>
<li><p><strong>fontsize</strong> (<em>int</em><em> or </em><em>str</em>) – Font size for labels and legend.</p></li>
<li><p><strong>colormap</strong> (<em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – List of colors to cycle through when signal.color is None.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>matplotlib.figure.Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal_collection.sum">
<span class="sig-name descname"><span class="pre">sum</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">indices_or_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coeffs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.signal_collection.sum" title="Link to this definition"></a></dt>
<dd><p>Sum selected signals, optionally weighted by coeffs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>indices_or_names</strong> (<em>list</em><em>[</em><em>int</em><em> or </em><em>str</em><em>]</em><em>, </em><em>optional</em>) – If provided, selects a subset by index or name.</p></li>
<li><p><strong>coeffs</strong> (<em>list</em><em>[</em><em>float</em><em>]</em><em>, </em><em>optional</em>) – Weights matching the number of selected signals.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Summed signal.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sig2dna_core.signomics.signal" title="sig2dna_core.signomics.signal">signal</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sig2dna_core.signomics.signal_collection.to_matrix">
<span class="sig-name descname"><span class="pre">to_matrix</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#sig2dna_core.signomics.signal_collection.to_matrix" title="Link to this definition"></a></dt>
<dd><p>Return a 2D array (n_signals x n_points) of aligned signal values.</p>
</dd></dl>

</section>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="README.html" class="btn btn-neutral float-left" title="📡🧬 sig2dna" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Olivier Vitrac (Generative Simulation).</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>